{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "53e89449-ba01-45ad-a186-62c734999ca7",
   "metadata": {},
   "source": [
    "# Notes on QC  \n",
    "Initially I filtered cells based on QC variables.  \n",
    "Then I reassessed the filtered cells and ran scrublet for doublet simulation.\n",
    "\n",
    "I will filter both with cutoffs and clusters. Sometimes it's difficult to design cutoffs that capture all the bad cells.  \n",
    "Clustering will often group them together well. If when assessing a sample I find a cluster of what I percieve as entirely bad cells, I will remove the cluster.  \n",
    "## CkD1\n",
    "Looking at CkD1 I chose the following filtering parameters:  \n",
    "'min_counts':7500, 'max_counts':100000, 'min_genes':3000, 'max_genes':15000,  'min_mito':2.3,  'max_mito':10,   'min_rb':7,  'max_rb':23  \n",
    "\n",
    "In our scatter and boxplots we can see 3 clusters that fall outside the filtering parameters.  \n",
    "Cluster 0, 12, 14 and 17  \n",
    "\n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD1_scatter_qc.png)  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD1_boxplot.png)  \n",
    "\n",
    "I want to check what they are before removing them.  \n",
    "\n",
    "They're all just poor quality cells of different kinds. Doesn't require much more investigation.  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD1_umap.png)  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD1_QC_umap.png)  \n",
    "\n",
    "\n",
    "I will mark the bad clusters as failed.:  \n",
    "['0','12','14','17']  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "Here cluster 36 stands out as being composed of likely doublets. In addition we can see small clumps of doublets around.  \n",
    "I filtered all cells by a doublet score of 0.1 and removed cluster 36.  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD1_pass_umap.png)   \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD1_pass_doublet2_umap.png)   \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD1_pass_highres_doublet_boxplot.png) \n",
    "\n",
    "## CkD2\n",
    "Looking at CkD2 I chose the following filtering parameters:  \n",
    "'min_counts':9000, 'max_counts':100000, 'min_genes':3000, 'max_genes':15000,  'min_mito':2.3,  'max_mito':10,   'min_rb':10,  'max_rb':24  \n",
    "\n",
    "In our scatter and boxplots we can see 8 clusters that fall outside the filtering parameters.  \n",
    "Cluster 0, 5, 12, '15',16, 18, 19, 26  \n",
    "\n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD2_scatter_qc.png)  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD2_boxplot.png)  \n",
    "\n",
    "I want to check what they are before removing them.  \n",
    "\n",
    "Only cluster 15 seems a little special. These seem to be vascular cells, it's a mix of low and good quality cells. I will keep this cluster and just filter based on cutoffs.  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD2_umap.png)  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD2_QC_umap.png)  \n",
    "\n",
    "\n",
    "I will mark the bad clusters as failed.:  \n",
    "['0','5','12','16','18','19','26']  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "Here cluster 20 stands out as being composed of likely doublets. In addition we can see small clumps of doublets around.  \n",
    "I filtered all cells by a doublet score of 0.15 and removed cluster 20.  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD2_pass_umap.png)   \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD2_pass_doublet2_umap.png)   \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD2_pass_doublet_boxplot.png) \n",
    "\n",
    "\n",
    "## CkD3\n",
    "Looking at CkD3 I chose the following filtering parameters:  \n",
    "'min_counts':6000, 'max_counts':100000, 'min_genes':2500, 'max_genes':15000,  'min_mito':1.3,  'max_mito':8,   'min_rb':8,  'max_rb':23  \n",
    "\n",
    "In our scatter and boxplots we can see 7 clusters that fall outside the filtering parameters.  \n",
    "Cluster 1, 4, 13, 18, 20, 24, 25  \n",
    "\n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD3_scatter_qc.png)  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD3_boxplot.png)  \n",
    "\n",
    "I want to check what they are before removing them.  \n",
    "Cluster 18 is vascular/endothelium, I won't outright remove it, just filter with params.  \n",
    "Cluster 20 is blood, I'll keept it for now as well.  \n",
    "Cluster 24 I think are immune cells or some other sort of hematopoetic derivative.  \n",
    "Cluster 25 looks like lens cells. Chicken embryos do have huge eyes after all.  \n",
    "\n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD3_umap.png)  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD3_QC_umap.png)  \n",
    "\n",
    "\n",
    "I will mark the bad clusters as failed.:  \n",
    "['1','4','13',]  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "Here cluster 8 stands out as being composed of likely doublets. In addition we can see small clumps of doublets around.  \n",
    "I filtered all cells by a doublet score of 0.18 and removed cluster 8.  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD3_pass_umap.png)   \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD3_pass_doublet2_umap.png)   \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD3_pass_doublet_boxplot.png) \n",
    "\n",
    "## CkD4\n",
    "Looking at CkD4 I chose the following filtering parameters:  \n",
    "This one is quite wild. There's a lot of cells and a lot of trash that overlaps with good cells. Hard to choose cutoffs. I will have to rely on clustering as well to filter out most of this.  \n",
    "'min_counts':4000, 'max_counts':100000, 'min_genes':2000, 'max_genes':15000,  'min_mito':1,  'max_mito':12.5,   'min_rb':9.5,  'max_rb':25  \n",
    "\n",
    "In our scatter and boxplots we can see 8 clusters that fall outside the filtering parameters.  \n",
    "Cluster 0, 1, 2, 4, 13, 15, 17, 19\n",
    "\n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD4_scatter_qc.png)  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD4_boxplot.png)  \n",
    "\n",
    "I want to check what they are before removing them.  \n",
    "Clusters 0, 1, 2 and 4 are trash.   \n",
    "Cluster 13 also looks like bad cells. No marker genes. But somehow different from the aforementioned 4 clusters.  \n",
    "Cluster 15 are endothelial cells\n",
    "Cluster 17 is blood\n",
    "Cluster 19 I think are immune cells or some other sort of hematopoetic derivative.  \n",
    "\n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD4_umap.png)  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD4_QC_umap.png)  \n",
    "\n",
    "\n",
    "I will mark the bad clusters as failed.:  \n",
    "['0', '1', '2', '4', '13',]  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "This sample was a bit trash, it might be harder to pick out the doublets here. For now I'll stick to the following but I might get back to it later.\n",
    "Here cluster 18 stands out as being composed of likely doublets. In addition we can see small clumps of doublets around.  \n",
    "I filtered all cells by a doublet score of 0.14 and removed cluster 18.  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD5_pass_umap.png)   \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD5_pass_doublet2_umap.png)   \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD5_pass_doublet_boxplot.png) \n",
    "\n",
    "## CkD5\n",
    "Looking at CkD5 I chose the following filtering parameters:  \n",
    "'min_counts':3700, 'max_counts':100000, 'min_genes':1700, 'max_genes':15000,  'min_mito':1.5,  'max_mito':10,   'min_rb':8,  'max_rb':27\n",
    "\n",
    "In our scatter and boxplots we can see 4 clusters that fall outside the filtering parameters.  \n",
    "Cluster 0, 13, 17, 23\n",
    "\n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD5_scatter_qc.png)  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD5_boxplot.png)  \n",
    "\n",
    "I want to check what they are before removing them.  \n",
    "Clusters 0 is trash.   \n",
    "Cluster 13, I'm not sure what it is, but I don't think it's trash.\n",
    "Cluster 17 is blood.  \n",
    "Cluster 23 I think are immune cells or some other sort of hematopoetic derivative.  \n",
    "\n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD5_umap.png)  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD5_QC_umap.png)  \n",
    "\n",
    "\n",
    "I will mark the bad clusters as failed.:  \n",
    "['0',]  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "This sample was a bit trash, it might be harder to pick out the doublets here. For now I'll stick to the following but I might get back to it later.\n",
    "Here clusters 12 and 16 stands out as being composed of likely doublets. In addition we can see a bunch of doublets with cluster 9, these never separate into a different cluster so I'll filter them by a cutoff.  \n",
    "I filtered all cells by a doublet score of 0.14 and removed clusters 12 and 16.  \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD5_pass_umap.png)   \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD5_pass_doublet2_umap.png)   \n",
    "![](../markdown_images/chicken/GRCg7b/markdown_plots/CkD5_pass_doublet_boxplot.png) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5d4584f-5bf6-444b-94cf-5c00459f2618",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.getLogger('matplotlib.font_manager').setLevel(logging.ERROR)\n",
    "import scanpy as sc\n",
    "import scvelo as scv\n",
    "import scvi\n",
    "import seaborn as sns\n",
    "import plotly.express as px\n",
    "import numpy as np\n",
    "from dash import Dash, dcc, html, Input, Output\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import warnings\n",
    "import logging\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "import gc\n",
    "os.environ['R_HOME'] = sys.exec_prefix+\"/lib/R/\"\n",
    "\n",
    "# Plotting\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "from matplotlib.colors import LinearSegmentedColormap, ListedColormap\n",
    "from matplotlib.lines import Line2D \n",
    "\n",
    "import logging\n",
    "logger = logging.getLogger()\n",
    "logger.setLevel(logging.CRITICAL)\n",
    "\n",
    "from copy import copy\n",
    "reds = copy(mpl.cm.Reds)\n",
    "reds.set_under(\"lightgray\")\n",
    "\n",
    "project_directory = '/Cranio_Lab/Louk_Seton/4_species_project'\n",
    "os.chdir(os.path.expanduser(\"~\")+project_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02349573-11ea-4c3f-b094-7c39b36e7d7a",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "##TODO\n",
    "\n",
    "######chicken timeeee######\n",
    "start_time=time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\")\n",
    "print('start time:',start_time)\n",
    "\n",
    "sample_names = ['CkD1','CkD2','CkD3','CkD4','CkD5',] #specify the sample names\n",
    "species = 'chicken' #specify the species\n",
    "genome = 'GRCg7b' #specify the genome\n",
    "output_prefix = 'cellranger_related/cellranger_output/sc_rnaseq/' #specify the location of the cellranger output\n",
    "\n",
    "mt_genes = ['ND1','ND2','ND3','ND4','ND4L','ND5','ND6','COX1','COX2','COX3','ATP6','ATP8','CYTB']\n",
    "rb_genes = ('RPS','RPL')\n",
    "hb_genes = ['HBBA','HBBR','HBAD','HBA1','HBZ','HBE','HBE1',]\n",
    "\n",
    "\n",
    "##plotting vars\n",
    "output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "genes_to_plot = ['SOX10',]\n",
    "dot_size = .6\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "##################################################################################################################\n",
    "cellranger_locs = []\n",
    "velocyto_locs = []\n",
    "\n",
    "for sample in sample_names: #simple loop to create a list with the location of the required files\n",
    "    cellranger_locs = cellranger_locs + [output_prefix+species+'/'+genome+'/'+sample+'/outs/filtered_feature_bc_matrix/']\n",
    "    velocyto_locs = velocyto_locs + [output_prefix+species+'/'+genome+'/'+sample+'/velocyto/'+sample+'.loom']\n",
    "\n",
    "warnings.filterwarnings('ignore')\n",
    "logging.getLogger().setLevel(logging.CRITICAL)\n",
    "\n",
    "adata_files={}\n",
    "print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Reading file(s)')\n",
    "for i,j,k in zip(sample_names, cellranger_locs, velocyto_locs):\n",
    "    ##Read and merge 10x and velocyto output, preserving gene names and ensemblid\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Sample ',i)\n",
    "    adata_cellr = sc.read_10x_mtx(j)\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Loaded 10x mtx')\n",
    "    adata_cellr.var['gene_name'] = adata_cellr.var.index\n",
    "    adata_cellr.var.index = adata_cellr.var['gene_ids']\n",
    "    adata_cellr.obs.index = adata_cellr.obs.index.str[:-2]\n",
    "   \n",
    "    adata_veloc = sc.read(k)\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Loaded velocyto loom')\n",
    "    adata_veloc.obs.index = adata_veloc.obs.index.str.split(':').str[1]\n",
    "    adata_veloc.obs.index = adata_veloc.obs.index.str[:-1]\n",
    "    adata_veloc.var.index = adata_veloc.var['Accession']\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Merging data')\n",
    "    adata_merge = scv.utils.merge(adata_cellr, adata_veloc)\n",
    "    adata_files[i]=adata_merge\n",
    "    \n",
    "    adata_files[i].obs['sample']=i\n",
    "    adata_files[i].obs['ref_genome']=genome\n",
    "    adata_files[i].obs['barcode']=adata_files[i].obs.index\n",
    "    adata_files[i].obs.index=adata_files[i].obs['barcode']+'.'+adata_files[i].obs['sample']\n",
    "\n",
    "    adata_files[i].var.index = list(adata_files[i].var['gene_name'])#set gene_names as index\n",
    "\n",
    "    ##Calculate QC values\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Calculating QC values for cells')\n",
    "    adata_files[i].var['mt'] = adata_files[i].var['gene_name'].isin(mt_genes)  # annotate the group of mitochondrial genes as 'mt'\n",
    "    adata_files[i].var['rb'] = adata_files[i].var['gene_name'].str.startswith(rb_genes)  # annotate the group of ribosomal proteins as 'rb'\n",
    "\n",
    "    adata_files[i].var['hb'] = adata_files[i].var['gene_name'].isin(hb_genes) #check this\n",
    "    adata_files[i].var_names_make_unique()\n",
    "    adata_files[i].obs_names_make_unique()\n",
    "    sc.pp.calculate_qc_metrics(adata_files[i], qc_vars=[\"mt\", \"rb\", \"hb\"], inplace=True, log1p=True)\n",
    "    #clip to meaningful numbers, better for comparisons\n",
    "    adata_files[i].obs['pct_counts_mt'] = adata_files[i].obs['pct_counts_mt'].clip(upper = 20)\n",
    "   # adata_files[i].obs['pct_counts_hb'] = adata_files[i].obs['pct_counts_hb'].clip(upper = 20)\n",
    "    #adata_files[i].obs['pct_counts_rb'] = adata_files[i].obs['pct_counts_rb'].clip(upper = 40)\n",
    "    adata_files[i].obs['total_counts'] = adata_files[i].obs['total_counts'].clip(upper = 50000)\n",
    "\n",
    "    ##calculate doublet score\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Running scrublet for doublet detection')\n",
    "    sc.pp.scrublet(adata_files[i])\n",
    "\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Initial preprocess and clustering')\n",
    "    ##standard normalization\n",
    "    adata_files[i].layers[\"original_counts\"] = adata_files[i].X.copy() #store original counts\n",
    "    sc.pp.normalize_total(adata_files[i]) # Normalizing to median total counts\n",
    "    sc.pp.log1p(adata_files[i]) # Logarithmize the data\n",
    "    adata_files[i].layers[\"normalized_counts\"] = adata_files[i].X.copy()\n",
    "\n",
    "    ##highly variable genes\n",
    "    sc.pp.highly_variable_genes(adata_files[i], n_top_genes=2000,)\n",
    "\n",
    "    ##dimensionality reduction and clustering\n",
    "    sc.tl.pca(adata_files[i])\n",
    "    sc.pp.neighbors(adata_files[i])\n",
    "    sc.tl.umap(adata_files[i])\n",
    "    sc.tl.leiden(adata_files[i],)\n",
    "\n",
    "    ##cell cycle scoring\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Prepare cell cycle scoring')\n",
    "    cell_cycle_genes = [x.strip() for x in open('required_files/chicken_cell_cycle.txt')]\n",
    "    s_genes = cell_cycle_genes[:43]\n",
    "    s_genes = [x for x in s_genes if x in list(adata_files[i].var['gene_name'])]\n",
    "    g2m_genes = cell_cycle_genes[43:]\n",
    "    g2m_genes = [x for x in g2m_genes if x in list(adata_files[i].var['gene_name'])]\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Cell cycle scoring')\n",
    "    sc.tl.score_genes_cell_cycle(adata_files[i], s_genes=s_genes, g2m_genes=g2m_genes)\n",
    "\n",
    "    \n",
    "    #plot pdfs of qc\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Plotting QC and saving to PDF')\n",
    "    with PdfPages(output_dir+i+'_QC.pdf') as pdf:\n",
    "        mpl.rcParams['axes.grid'] = False\n",
    "        ax = sc.pl.violin(adata_files[i],[\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\",\"pct_counts_hb\", \"pct_counts_rb\",'doublet_score'],\n",
    "                          stripplot=False, multi_panel=True,show=False,\n",
    "                          height=4, aspect=.8,col_wrap = 3)\n",
    "        plt.suptitle(i, x = 0.1)\n",
    "        pdf.savefig()  # saves the current figure into a pdf page\n",
    "        plt.close() # close the plot and page\n",
    "\n",
    "        ###Scatterplots\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap=scatter_cmap\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files[i].obs, x=x1,y=y1,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax1.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax1.set_ylabel(y1)\n",
    "        ax2.scatter(data=adata_files[i].obs, x=x2,y=y1,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax2.set_xlim(adata_files[i].obs[x2].quantile(0) ,adata_files[i].obs[x2].quantile(0.95))\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax2.set_ylabel(y1)\n",
    "        ax3.scatter(data=adata_files[i].obs, x=x1,y=y2,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax3.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "        ax3.set_xlabel(x1) \n",
    "        ax3.set_ylabel(y2)\n",
    "        ax4.scatter(data=adata_files[i].obs, x=x2,y=y2,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax4.set_xlim(adata_files[i].obs[x2].quantile(0) ,adata_files[i].obs[x2].quantile(0.95))\n",
    "        ax4.set_xlabel(x2) \n",
    "        ax4.set_ylabel(y2)\n",
    "\n",
    "\n",
    "        fig.colorbar(mpl.cm.ScalarMappable(norm=mpl.colors.Normalize(0, max(adata_files[i].obs[c])), cmap=cmap),\n",
    "                     ax=ax1, orientation='vertical', label=c)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##nice jointplot shows good cells\n",
    "        mpl.rcParams['axes.grid'] = False\n",
    "        sns_plot = sns.jointplot(data=adata_files[i].obs,\n",
    "                                 x=\"log1p_total_counts\",\n",
    "                                 y=\"log1p_n_genes_by_counts\",\n",
    "                                 kind=\"hex\",\n",
    "                                 marginal_kws=dict(bins=100,),\n",
    "                                 joint_kws = dict(gridsize=50,),\n",
    "                                )\n",
    "        sns_plot.fig.suptitle(i,x=0.05)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##custom hvg plot\n",
    "        fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize = (8,4))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap = ListedColormap(['grey','black'])\n",
    "        c = 'highly_variable'\n",
    "        scatter = ax1.scatter(data=adata_files[i].var, x='means',y=\"dispersions\",\n",
    "                              c=c,s = s,cmap=cmap,linewidth=linewidth,rasterized=True)\n",
    "        ax1.set_xlabel('mean expression') \n",
    "        ax1.set_ylabel('dispersion') \n",
    "        ax1.legend(scatter.legend_elements()[0][0:2],['other genes','highly variable'],\n",
    "                   loc=\"upper right\",)\n",
    "        scatter = ax2.scatter(data=adata_files[i].var, x='means',y=\"dispersions_norm\",\n",
    "                              c=c,s = s,cmap=cmap,linewidth=linewidth,rasterized=True)\n",
    "        ax2.legend(scatter.legend_elements()[0][0:2],['other genes','highly variable'],\n",
    "                   loc=\"upper right\",)\n",
    "        ax2.set_xlabel('mean expression') \n",
    "        ax2.set_ylabel('dispersions_norm')\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##PCA variance plot\n",
    "        sc.pl.pca_variance_ratio(adata_files[i], log=True, show=False)\n",
    "        pdf.savefig()\n",
    "        plt.close()\n",
    "        \n",
    "        ##umaps\n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files[i],\n",
    "                        color=['doublet_score','leiden','total_counts','pct_counts_mt','pct_counts_rb','pct_counts_hb','phase',]+genes_to_plot,\n",
    "                        show = False,ncols = 2)\n",
    "        for p in ax:\n",
    "            p.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##Most highly expressed genes\n",
    "        plt.rcParams['figure.figsize'] = [8,4]\n",
    "        adata_files[i].X = adata_files[i].layers[\"original_counts\"].copy() #set raw counts to X because the layer argument is broken\n",
    "        ax = sc.pl.highest_expr_genes(adata_files[i], \n",
    "                                      gene_symbols = 'gene_name',\n",
    "                                      n_top=20,log=True,show=False)\n",
    "        ax.set_rasterized(True)\n",
    "        pdf.savefig()\n",
    "        plt.close()\n",
    "        ##return normalised counts to X\n",
    "        adata_files[i].X = adata_files[i].layers[\"normalized_counts\"].copy()\n",
    "\n",
    "        ##doublet scatterplots\n",
    "        ###Scatterplots\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap=scatter_cmap\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files[i].obs, x=x1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax2.scatter(data=adata_files[i].obs, x=x2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax3.scatter(data=adata_files[i].obs, x=y1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax3.set_xlabel(y1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax4.scatter(data=adata_files[i].obs, x=y2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax4.set_xlabel(y2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        fig.colorbar(mpl.cm.ScalarMappable(norm=mpl.colors.Normalize(0, max(adata_files[i].obs[c])), cmap=cmap),\n",
    "                     ax=ax1, orientation='vertical', label=c)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "        \n",
    "        #boxplot of doublet scores\n",
    "        ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"doublet_score\",\n",
    "            palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0)\n",
    "        cnum = 0\n",
    "        for line in ax.lines:\n",
    "            if line.get_linestyle() == 'None':\n",
    "                sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                              size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True)\n",
    "                cnum = cnum+1\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "    ##saving h5ad\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Saving data to h5ad file')\n",
    "    !mkdir -p {'h5ad_files/'+species+'/'+genome}\n",
    "    adata_files[i].write('h5ad_files/'+species+'/'+genome+'/'+i+'_before_filtering.h5ad')\n",
    "    \n",
    "\n",
    "    ##remove used variables from memory\n",
    "    del adata_cellr\n",
    "    del adata_veloc\n",
    "    del adata_merge\n",
    "    gc.collect()\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Loaded files for ',i)\n",
    "print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Finished Reading file(s)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c09be03b-1772-440b-9bbe-40d9f232b4d7",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "##Filter using QC vars and make plots\n",
    "filter_vars = {'CkD1':{'min_counts':7500, 'max_counts':100000, 'min_genes':3000, 'max_genes':15000,  'min_mito':2.3,  'max_mito':10,   'min_rb':7,  'max_rb':23,\n",
    "                      'clusters':['0','12','14','17']},\n",
    "               'CkD2':{'min_counts':9000, 'max_counts':100000, 'min_genes':3000, 'max_genes':15000,  'min_mito':2.3,  'max_mito':10,   'min_rb':10,  'max_rb':24,\n",
    "                      'clusters':['0','5','12','16','18','19','26']}, \n",
    "               'CkD3':{'min_counts':6000, 'max_counts':100000, 'min_genes':2500, 'max_genes':15000,  'min_mito':1.3,  'max_mito':8,   'min_rb':8,  'max_rb':23,\n",
    "                      'clusters':['1','4','13',]},\n",
    "               'CkD4':{'min_counts':4000, 'max_counts':100000, 'min_genes':2000, 'max_genes':15000,  'min_mito':1,  'max_mito':12.5,   'min_rb':9.5,  'max_rb':25,\n",
    "                      'clusters':['0', '1', '2', '4', '13',]},\n",
    "               'CkD5':{'min_counts':3700, 'max_counts':100000, 'min_genes':1700, 'max_genes':15000,  'min_mito':1.5,  'max_mito':10,   'min_rb':8,  'max_rb':27,\n",
    "                      'clusters':['0']},}\n",
    "\n",
    "filter_vars = pd.DataFrame(filter_vars).T\n",
    "\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "dot_size = 1\n",
    "linewidth = 0\n",
    "scatter_color = 'QC'\n",
    "output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "for sample in list(filter_vars.index):\n",
    "    min_counts = filter_vars.loc[sample,'min_counts']\n",
    "    max_counts = filter_vars.loc[sample,'max_counts']\n",
    "    min_genes = filter_vars.loc[sample,'min_genes']\n",
    "    max_genes = filter_vars.loc[sample,'max_genes']\n",
    "    min_mito = filter_vars.loc[sample,'min_mito']\n",
    "    max_mito = filter_vars.loc[sample,'max_mito']\n",
    "    min_rb = filter_vars.loc[sample,'min_rb']\n",
    "    max_rb = filter_vars.loc[sample,'max_rb']\n",
    "\n",
    "    conditions = [\n",
    "    (adata_files[sample].obs['n_genes_by_counts'] < min_genes),\n",
    "    (adata_files[sample].obs['n_genes_by_counts'] > max_genes),\n",
    "    (adata_files[sample].obs['total_counts'] < min_counts),\n",
    "    (adata_files[sample].obs['total_counts'] > max_counts),\n",
    "    (adata_files[sample].obs['pct_counts_mt'] > max_mito),\n",
    "    (adata_files[sample].obs['pct_counts_mt'] < min_mito),\n",
    "    (adata_files[sample].obs['pct_counts_rb'] < min_rb),\n",
    "    (adata_files[sample].obs['pct_counts_rb'] > max_rb),\n",
    "    (adata_files[sample].obs['pct_counts_mt'] <= max_mito) & (adata_files[sample].obs['pct_counts_mt'] >= min_mito) & (adata_files[sample].obs['n_genes_by_counts'] >= min_genes) & (adata_files[sample].obs['n_genes_by_counts'] <= max_genes) & \n",
    "    (adata_files[sample].obs['total_counts'] >= min_counts) & (adata_files[sample].obs['total_counts'] <= max_counts) & (adata_files[sample].obs['pct_counts_rb'] >= min_rb) & (adata_files[sample].obs['pct_counts_rb'] <= max_rb)\n",
    "    ]\n",
    "    \n",
    "    values = ['Low_nFeature', 'High_nFeature', 'Low_Counts','High_Counts', 'High_MT','Low_MT', 'Low_RB' ,'High_RB' ,'Pass']\n",
    "    adata_files[sample].obs['QC'] = np.select(conditions, values)\n",
    "    adata_files[sample].obs.loc[adata_files[sample].obs['leiden'].isin(filter_vars.loc[sample,'clusters']), 'QC'] = \"Bad_Cluster\"\n",
    "    adata_files[sample].obs['QC'] = adata_files[sample].obs['QC'].astype('category')\n",
    "\n",
    "    colors = {'Low_nFeature':'#d9d9d9', \n",
    "                  'High_nFeature':'#8dd3c7', \n",
    "                  'Low_Counts':'#ffffb3',\n",
    "                  'High_Counts':'#bebada',\n",
    "                  'High_MT':'#fb8072',\n",
    "                  'Low_MT':'#80b1d3',\n",
    "                  'Low_RB':'#fdb462',\n",
    "                  'High_RB':'#fccde5',\n",
    "                  'Bad_Cluster':'#000000',\n",
    "                  'Pass':'#b3de69',}\n",
    "    with PdfPages(output_dir+sample+'_filter_params.pdf') as pdf:        \n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files[sample], color = ['QC'],palette = colors, show=False)\n",
    "        ax.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150,bbox_inches='tight')  # saves the current figure into a pdf page\n",
    "        plt.close() # close the plot and page\n",
    "        \n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files[sample], color = ['n_genes_by_counts','doublet_score','phase','pct_counts_hb'],ncols=2,show=False)\n",
    "        for p in ax:\n",
    "            p.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150)  # saves the current figure into a pdf page\n",
    "        plt.close() # close the plot and page\n",
    "\n",
    "        ###Scatterplots\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files[sample].obs, x=x1,y=y1,\n",
    "                    c = adata_files[sample].obs[c].map(colors),s = s,linewidth=linewidth,rasterized=True)\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax1.set_ylabel(y1)\n",
    "        ax2.scatter(data=adata_files[sample].obs, x=x2,y=y1,\n",
    "                    c = adata_files[sample].obs[c].map(colors),s = s,linewidth=linewidth,rasterized=True)\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax2.set_ylabel(y1)\n",
    "        ax3.scatter(data=adata_files[sample].obs, x=x1,y=y2,\n",
    "                    c = adata_files[sample].obs[c].map(colors),s = s,linewidth=linewidth,rasterized=True)\n",
    "        ax3.set_xlabel(x1) \n",
    "        ax3.set_ylabel(y2)\n",
    "        scatter = ax4.scatter(data=adata_files[sample].obs, x=x2,y=y2,\n",
    "                    c = adata_files[sample].obs[c].map(colors),s = s,linewidth=linewidth,rasterized=True)\n",
    "        ax4.set_xlabel(x2) \n",
    "        ax4.set_ylabel(y2)\n",
    "        handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=8) for k, v in colors.items()]\n",
    "        ax4.legend(title='color', handles=handles, bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7b4a01f-93b5-4ba5-b841-552290bf079a",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "##doublet stuff and some more plots\n",
    "genes_to_plot = ['SOX10','SHH','SOX21','PAX7','HBA1','TWIST2']\n",
    "dot_size = 1\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "\n",
    "for i in sample_names:\n",
    "    adata_files[i+'_pass'] = adata_files[i][adata_files[i].obs['QC']=='Pass'].copy() #filter cells\n",
    "    i = i+'_pass'\n",
    "    adata_files[i].X = adata_files[i].layers[\"original_counts\"].copy() #restore original counts\n",
    "    sc.pp.scrublet(adata_files[i])\n",
    "\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Initial preprocess and clustering')\n",
    "    ##standard normalization\n",
    "    sc.pp.normalize_total(adata_files[i]) # Normalizing to median total counts\n",
    "    sc.pp.log1p(adata_files[i]) # Logarithmize the data\n",
    "    adata_files[i].layers[\"normalized_counts\"] = adata_files[i].X.copy()\n",
    "\n",
    "    ##highly variable genes\n",
    "    sc.pp.highly_variable_genes(adata_files[i], n_top_genes=2000,)\n",
    "\n",
    "    ##dimensionality reduction and clustering\n",
    "    sc.tl.pca(adata_files[i])\n",
    "    sc.pp.neighbors(adata_files[i])\n",
    "    sc.tl.umap(adata_files[i])\n",
    "    sc.tl.leiden(adata_files[i],resolution = 1, key_added = 'leiden_filt')\n",
    "    sc.tl.leiden(adata_files[i],resolution = 2, key_added = 'leiden_filt_highres')\n",
    "\n",
    "    ##plotting new stuff\n",
    "    output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "    !mkdir -p {output_dir}\n",
    "\n",
    "    with PdfPages(output_dir+i+'_doublet_check.pdf') as pdf:\n",
    "        ##umaps\n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files[i],\n",
    "                        color=['doublet_score','leiden','total_counts','leiden_filt','pct_counts_mt','leiden_filt_highres','pct_counts_rb','pct_counts_hb','phase',]+genes_to_plot,\n",
    "                        show = False,ncols = 2)\n",
    "        for p in ax:\n",
    "            p.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "        \n",
    "        #boxplot of doublet scores\n",
    "        plt.rcParams['figure.figsize'] = [8,4]\n",
    "        ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"doublet_score\",\n",
    "            palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "            palette = adata_files[i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "        cnum = 0\n",
    "        for line in ax.lines:\n",
    "            if line.get_linestyle() == 'None':\n",
    "                sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                              size = 1.5, color = adata_files[i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "                cnum = cnum+1\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt_highres\", y=\"doublet_score\",\n",
    "            palette = adata_files[i].uns['leiden_filt_highres_colors'], \n",
    "                 fliersize=0)\n",
    "        cnum = 0\n",
    "        for line in ax.lines:\n",
    "            if line.get_linestyle() == 'None':\n",
    "                sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                              size = 1.5, color = adata_files[i].uns['leiden_filt_highres_colors'][cnum], rasterized = True)\n",
    "                cnum = cnum+1\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##doublet scatterplots\n",
    "        ###Scatterplots\n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap=scatter_cmap\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files[i].obs, x=x1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax2.scatter(data=adata_files[i].obs, x=x2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax3.scatter(data=adata_files[i].obs, x=y1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax3.set_xlabel(y1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax4.scatter(data=adata_files[i].obs, x=y2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax4.set_xlabel(y2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        fig.colorbar(mpl.cm.ScalarMappable(norm=mpl.colors.Normalize(0, max(adata_files[i].obs[c])), cmap=cmap),\n",
    "                     ax=ax1, orientation='vertical', label=c)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06a6c877-428d-4528-81a6-f6e5348d521c",
   "metadata": {},
   "outputs": [],
   "source": [
    "##removing doublets and generating data for analysis\n",
    "doublet_vars = {'CkD1_pass':{'doublet_cutoff':.1, 'doublet_clusters':None,'doublet_clusters_high':['36']},\n",
    "                'CkD2_pass':{'doublet_cutoff':.15, 'doublet_clusters':['20'],'doublet_clusters_high':None},\n",
    "                'CkD3_pass':{'doublet_cutoff':.18, 'doublet_clusters':['8'],'doublet_clusters_high':None},\n",
    "                'CkD4_pass':{'doublet_cutoff':.14, 'doublet_clusters':['18'],'doublet_clusters_high':None},\n",
    "                'CkD5_pass':{'doublet_cutoff':.14, 'doublet_clusters':['12','16',],'doublet_clusters_high':None},\n",
    "               }\n",
    "\n",
    "doublet_vars = pd.DataFrame(doublet_vars).T\n",
    "\n",
    "genes_to_plot = ['SHH','TWIST2','HBA1','SOX10','SOX21']\n",
    "\n",
    "for sample in list(doublet_vars.index):\n",
    "    if 'pass' in sample:\n",
    "        doublet_cutoff = doublet_vars.loc[sample,'doublet_cutoff']\n",
    "        doublet_clusters = doublet_vars.loc[sample,'doublet_clusters']\n",
    "        doublet_clusters_high = doublet_vars.loc[sample,'doublet_clusters_high']\n",
    "        adata_files[sample].obs['doublet_cutoff'] = 'pass'\n",
    "        adata_files[sample].obs.loc[adata_files[sample].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "        if isinstance(doublet_clusters, list):\n",
    "            adata_files[sample].obs.loc[adata_files[sample].obs['leiden_filt'].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "        if isinstance(doublet_clusters_high, list):\n",
    "            adata_files[sample].obs.loc[adata_files[sample].obs['leiden_filt_highres'].isin(doublet_clusters_high), 'doublet_cutoff'] = 'doublet'\n",
    "        adata_files[sample].obs['doublet_cutoff'] = adata_files[sample].obs['doublet_cutoff'].astype('category')\n",
    "\n",
    "        sn = sample+'_doublet'\n",
    "        adata_files[sn] = adata_files[sample][adata_files[sample].obs['doublet_cutoff'] == 'pass'].copy()\n",
    "        adata_files[sn].X = adata_files[sn].layers[\"original_counts\"].copy() #restore original counts\n",
    "        print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Initial preprocess and clustering')\n",
    "        ##standard normalization\n",
    "        sc.pp.normalize_total(adata_files[sn]) # Normalizing to median total counts\n",
    "        sc.pp.log1p(adata_files[sn]) # Logarithmize the data\n",
    "        adata_files[sn].layers[\"normalized_counts\"] = adata_files[sn].X.copy()\n",
    "\n",
    "        ##highly variable genes\n",
    "        sc.pp.highly_variable_genes(adata_files[sn], n_top_genes=1000,)\n",
    "\n",
    "        ##dimensionality reduction and clustering\n",
    "        sc.tl.pca(adata_files[sn])\n",
    "        sc.pp.neighbors(adata_files[sn])\n",
    "        sc.tl.umap(adata_files[sn])\n",
    "        sc.tl.leiden(adata_files[sn],resolution = 1, key_added = 'leiden_post_QC')\n",
    "        sc.tl.leiden(adata_files[sn],resolution = 2, key_added = 'leiden_post_QC_highres')\n",
    "\n",
    "        ##plotting new stuff\n",
    "        output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "        !mkdir -p {output_dir}\n",
    "\n",
    "        i = sample\n",
    "        with PdfPages(output_dir+i+'_post_QC.pdf') as pdf:\n",
    "            ##umaps\n",
    "            plt.rcParams['figure.figsize'] = [5,4]\n",
    "            ax = sc.pl.umap(adata_files[sn],\n",
    "                            color=['leiden_post_QC','leiden_post_QC_highres','phase',]+genes_to_plot,\n",
    "                            show = False,ncols = 2)\n",
    "            for p in ax:\n",
    "                p.set_rasterized(True)\n",
    "            pdf.savefig(dpi=150)\n",
    "            plt.close()\n",
    "        \n",
    "        adata_files[sn].write('h5ad_files/'+species+'/'+genome+'/'+i+'_after_filtering.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03972365-cd19-483d-8131-6df12459a597",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "##markdown plots\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "## let's start with CkD1\n",
    "i = 'CkD1'\n",
    "filter_vars = {'CkD1':{'min_counts':7500, 'max_counts':100000, 'min_genes':3000, 'max_genes':15000,  'min_mito':2.3,  'max_mito':10,   'min_rb':7,  'max_rb':23}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files[i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files[i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files[i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,12)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files[i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax2.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax2.set_ylim(0,28)\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplots\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden','total_counts','pct_counts_mt',],groups = ['0','12','14','17'], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "##markdown plots\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "## let's start with CkD2\n",
    "i = 'CkD2'\n",
    "filter_vars = {'CkD2':{'min_counts':9000, 'max_counts':100000, 'min_genes':3000, 'max_genes':15000,  'min_mito':2.3,  'max_mito':10,   'min_rb':10,  'max_rb':24}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files[i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files[i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files[i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,12)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files[i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax2.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax2.set_ylim(0,28)\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplots\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden','total_counts','pct_counts_mt','CDH5'],groups = ['0','5','12','16','18','19','26'], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "##markdown plots\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "## let's start with CkD3\n",
    "i = 'CkD3'\n",
    "filter_vars = {'CkD3':{'min_counts':6000, 'max_counts':100000, 'min_genes':2500, 'max_genes':15000,  'min_mito':1.3,  'max_mito':8,   'min_rb':8,  'max_rb':23}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files[i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files[i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files[i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,12)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files[i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax2.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax2.set_ylim(0,28)\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplots\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden','total_counts','pct_counts_mt','CDH5','HBA1','CTSS','GJA3',],groups = ['1','4','13',], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "##markdown plots\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "## let's start with CkD4\n",
    "i = 'CkD4'\n",
    "filter_vars = {'CkD4':{'min_counts':4000, 'max_counts':100000, 'min_genes':2000, 'max_genes':15000,  'min_mito':1,  'max_mito':12.5,   'min_rb':9.5,  'max_rb':25}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files[i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files[i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files[i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,12)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files[i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax2.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax2.set_ylim(0,28)\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplots\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden','total_counts','pct_counts_mt','CDH5','HBA1','CTSS',],groups = ['0', '1', '2', '4', '13',], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "##markdown plots\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "## let's start with CkD5\n",
    "i = 'CkD5'\n",
    "filter_vars = {'CkD5':{'min_counts':3700, 'max_counts':100000, 'min_genes':1700, 'max_genes':15000,  'min_mito':1.5,  'max_mito':10,   'min_rb':8,  'max_rb':27}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files[i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files[i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files[i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,12)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files[i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax2.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax2.set_ylim(0,28)\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplots\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden','total_counts','pct_counts_mt','HBA1','CTSS',],groups = ['0',], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "## post filter plots\n",
    "filter_vars = {'CkD1':{'min_counts':7500, 'max_counts':100000, 'min_genes':3000, 'max_genes':15000,  'min_mito':2.3,  'max_mito':10,   'min_rb':7,  'max_rb':23,\n",
    "                      'clusters':['0','12','14','17']},\n",
    "               'CkD2':{'min_counts':9000, 'max_counts':100000, 'min_genes':3000, 'max_genes':15000,  'min_mito':2.3,  'max_mito':10,   'min_rb':10,  'max_rb':24,\n",
    "                      'clusters':['0','5','12','16','18','19','26']}, \n",
    "               'CkD3':{'min_counts':6000, 'max_counts':100000, 'min_genes':2500, 'max_genes':15000,  'min_mito':1.3,  'max_mito':8,   'min_rb':8,  'max_rb':23,\n",
    "                      'clusters':['1','4','13',]},\n",
    "               'CkD4':{'min_counts':4000, 'max_counts':100000, 'min_genes':2000, 'max_genes':15000,  'min_mito':1,  'max_mito':12.5,   'min_rb':9.5,  'max_rb':25,\n",
    "                      'clusters':['0', '1', '2', '4', '13',]},\n",
    "               'CkD5':{'min_counts':3700, 'max_counts':100000, 'min_genes':1700, 'max_genes':15000,  'min_mito':1.5,  'max_mito':10,   'min_rb':8,  'max_rb':27,\n",
    "                      'clusters':['0']},}\n",
    "\n",
    "filter_vars = pd.DataFrame(filter_vars).T\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "for sample in list(filter_vars.index):\n",
    "    min_counts = filter_vars.loc[sample,'min_counts']\n",
    "    max_counts = filter_vars.loc[sample,'max_counts']\n",
    "    min_genes = filter_vars.loc[sample,'min_genes']\n",
    "    max_genes = filter_vars.loc[sample,'max_genes']\n",
    "    min_mito = filter_vars.loc[sample,'min_mito']\n",
    "    max_mito = filter_vars.loc[sample,'max_mito']\n",
    "    min_rb = filter_vars.loc[sample,'min_rb']\n",
    "    max_rb = filter_vars.loc[sample,'max_rb']\n",
    "\n",
    "    conditions = [\n",
    "    (adata_files[sample].obs['n_genes_by_counts'] < min_genes),\n",
    "    (adata_files[sample].obs['n_genes_by_counts'] > max_genes),\n",
    "    (adata_files[sample].obs['total_counts'] < min_counts),\n",
    "    (adata_files[sample].obs['total_counts'] > max_counts),\n",
    "    (adata_files[sample].obs['pct_counts_mt'] > max_mito),\n",
    "    (adata_files[sample].obs['pct_counts_mt'] < min_mito),\n",
    "    (adata_files[sample].obs['pct_counts_rb'] < min_rb),\n",
    "    (adata_files[sample].obs['pct_counts_rb'] > max_rb),\n",
    "    (adata_files[sample].obs['pct_counts_mt'] <= max_mito) & (adata_files[sample].obs['pct_counts_mt'] >= min_mito) & (adata_files[sample].obs['n_genes_by_counts'] >= min_genes) & (adata_files[sample].obs['n_genes_by_counts'] <= max_genes) & \n",
    "    (adata_files[sample].obs['total_counts'] >= min_counts) & (adata_files[sample].obs['total_counts'] <= max_counts) & (adata_files[sample].obs['pct_counts_rb'] >= min_rb) & (adata_files[sample].obs['pct_counts_rb'] <= max_rb)\n",
    "    ]\n",
    "    \n",
    "    values = ['Fail', 'Fail', 'Fail','Fail', 'Fail','Fail', 'Fail' ,'Fail' ,'Pass']\n",
    "    adata_files[sample].obs['QC'] = np.select(conditions, values)\n",
    "    adata_files[sample].obs.loc[adata_files[sample].obs['leiden'].isin(filter_vars.loc[sample,'clusters']), 'QC'] = \"Fail\"\n",
    "    adata_files[sample].obs['QC'] = adata_files[sample].obs['QC'].astype('category')\n",
    "\n",
    "    plt.rcParams['figure.figsize'] = [4,3]\n",
    "    ax = sc.pl.umap(adata_files[sample], color = ['QC',], ncols = 3,show = False)\n",
    "    plt.savefig(output_dir+sample+'_QC_umap.png', dpi = 80,bbox_inches='tight')\n",
    "    plt.close()\n",
    "\n",
    "####Doublet stuff####\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "species = 'chicken' #specify the species\n",
    "genome = 'GRCg7b' #specify the genome\n",
    "#CkD1\n",
    "i = 'CkD1_pass'\n",
    "doublet_cutoff = .1\n",
    "doublet_clusters = ['36']\n",
    "doublet_filt_clust = 'leiden_filt_highres'\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden_filt','leiden_filt_highres','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = doublet_cutoff,vmax = .4)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "##################################################################\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt_highres\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_highres_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_highres_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_highres_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files[i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files[i].obs.loc[adata_files[i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs.loc[adata_files[i].obs[doublet_filt_clust].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs['doublet_cutoff'] = adata_files[i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], size = 5,color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = 'doublet_cutoff', groups = ['doublet'],show = False)\n",
    "plt.savefig(output_dir+i+'_doublet2_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "####Doublet stuff####\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "species = 'chicken' #specify the species\n",
    "genome = 'GRCg7b' #specify the genome\n",
    "#CkD1\n",
    "i = 'CkD2_pass'\n",
    "doublet_cutoff = .15\n",
    "doublet_clusters = ['20']\n",
    "doublet_filt_clust = 'leiden_filt'\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden_filt','leiden_filt_highres','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = doublet_cutoff,vmax = .4)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "##################################################################\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt_highres\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_highres_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_highres_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_highres_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files[i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files[i].obs.loc[adata_files[i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs.loc[adata_files[i].obs[doublet_filt_clust].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs['doublet_cutoff'] = adata_files[i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], size = 5,color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = 'doublet_cutoff', groups = ['doublet'],show = False)\n",
    "plt.savefig(output_dir+i+'_doublet2_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "####Doublet stuff####\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "species = 'chicken' #specify the species\n",
    "genome = 'GRCg7b' #specify the genome\n",
    "#CkD3\n",
    "i = 'CkD3_pass'\n",
    "doublet_cutoff = .18\n",
    "doublet_clusters = ['8']\n",
    "doublet_filt_clust = 'leiden_filt'\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden_filt','leiden_filt_highres','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = doublet_cutoff,vmax = .4)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "##################################################################\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt_highres\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_highres_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_highres_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_highres_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files[i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files[i].obs.loc[adata_files[i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs.loc[adata_files[i].obs[doublet_filt_clust].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs['doublet_cutoff'] = adata_files[i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], size = 5,color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = 'doublet_cutoff', groups = ['doublet'],show = False)\n",
    "plt.savefig(output_dir+i+'_doublet2_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "####Doublet stuff####\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "species = 'chicken' #specify the species\n",
    "genome = 'GRCg7b' #specify the genome\n",
    "#CkD4\n",
    "i = 'CkD4_pass'\n",
    "doublet_cutoff = .14\n",
    "doublet_clusters = ['18']\n",
    "doublet_filt_clust = 'leiden_filt'\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden_filt','leiden_filt_highres','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = doublet_cutoff,vmax = .4)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "##################################################################\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt_highres\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_highres_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_highres_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_highres_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files[i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files[i].obs.loc[adata_files[i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs.loc[adata_files[i].obs[doublet_filt_clust].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs['doublet_cutoff'] = adata_files[i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], size = 5,color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = 'doublet_cutoff', groups = ['doublet'],show = False)\n",
    "plt.savefig(output_dir+i+'_doublet2_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "####Doublet stuff####\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "species = 'chicken' #specify the species\n",
    "genome = 'GRCg7b' #specify the genome\n",
    "#CkD5\n",
    "i = 'CkD5_pass'\n",
    "doublet_cutoff = .14\n",
    "doublet_clusters = ['12','16']\n",
    "doublet_filt_clust = 'leiden_filt'\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden_filt','leiden_filt_highres','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = doublet_cutoff,vmax = .4)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "##################################################################\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt_highres\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_highres_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_highres_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_highres_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files[i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files[i].obs.loc[adata_files[i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs.loc[adata_files[i].obs[doublet_filt_clust].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs['doublet_cutoff'] = adata_files[i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], size = 5,color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = 'doublet_cutoff', groups = ['doublet'],show = False)\n",
    "plt.savefig(output_dir+i+'_doublet2_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a3c015f-e6d9-4f6e-bdb4-1772f8b2dc32",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf71f30f-7a56-4f1f-9b63-421f6c9b4412",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata_files[i], 'leiden_filt', method = 'wilcoxon')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cff6dee-47bb-49a1-aac9-f24eb3c396b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    adata_files[i],\n",
    "    groups = ['20'],\n",
    "    n_genes=5,\n",
    "    values_to_plot=\"logfoldchanges\", cmap='bwr',\n",
    "    vmin=-4,\n",
    "    vmax=4,\n",
    "    min_logfoldchange=2,\n",
    "    colorbar_title='log fold change'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12dcfe96-8749-4ac9-8b5b-0a4291ef5584",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_files[i], color = ['leiden','EPCAM','RFX4','PAX7'], groups = ['13'], cmap = reds, vmin = 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e83f6e8-7145-43f9-84ab-943c40076e6d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy_2025",
   "language": "python",
   "name": "scanpy_2025"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
