{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "23c419ac-a53a-4c51-ada3-467030485b4b",
   "metadata": {},
   "source": [
    "# Notes on QC  \n",
    "Initially I filtered cells based on QC variables.  \n",
    "Then I reassessed the filtered cells and ran scrublet for doublet simulation.\n",
    "\n",
    "I will filter both with cutoffs and clusters. Sometimes it's difficult to design cutoffs that capture all the bad cells.  \n",
    "Clustering will often group them together well. If when assessing a sample I find a cluster of what I percieve as entirely bad cells, I will remove the cluster.  \n",
    "## ME8\n",
    "Looking at ME8 I chose the following filtering parameters:  \n",
    "'min_counts':10000, 'max_counts':100000, 'min_genes':2900, 'max_genes':10000,  'min_mito':.1,  'max_mito':1.5,   'min_rb':15,  'max_rb':31  \n",
    "\n",
    "In our scatter and boxplots we can see 5 clusters that fall outside the filtering parameters.  \n",
    "Clusters 12 and 15 -> high Rb, low Mt and low genes \n",
    "Cluster 18 -> low genes \n",
    "Cluster 20 -> low Rb, low Mt and low genes  \n",
    "Cluster 22 -> low Mt and low genes  \n",
    "\n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME8_scatter_qc.png)  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME8_boxplot.png)  \n",
    "\n",
    "I want to check what they are before removing them.  \n",
    "\n",
    "Cluster 12 -> Poor quality brain cells. Very high Rb and low Mt. (Pou3f1)  \n",
    "Cluster 15 -> Poor quality ectoderm cells. Very high Rb and low Mt. (Epcam)  \n",
    "Cluster 18 -> Poor quality cells. No markers at all. Low counts  \n",
    "Cluster 20 -> Blood cells. Hence the low transcripts and low Rb and Mt  \n",
    "Cluster 22 -> Poor quality muscle cells. absent mitochondrial transcripts. Likely apoptotic or only nucleus. (Myl4)  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME8_umap.png)  \n",
    "\n",
    "I will mark the bad clusters as failed.:  \n",
    "['12','15','18','20','22']\n",
    "\n",
    "Cells after QC:  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME8_QC_umap.png)  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "Here cluster 14 stands out as being composed of likely doublets. In addition we can see small clumps of doublets around.  \n",
    "I filtered all cells by a doublet score of 0.15 and removed cluster 14.  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME8_pass_umap.png) ![](../markdown_images/mouse/mm10/markdown_plots/ME8_pass_doublet_umap.png)   \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME8_pass_doublet_boxplot.png)   \n",
    "\n",
    "\n",
    "\n",
    "## ME9\n",
    "At ME9 I chose the following filtering parameters:  \n",
    "'min_counts':9000, 'max_counts':100000, 'min_genes':4000, 'max_genes':10000,  'min_mito':.2,  'max_mito':2,   'min_rb':10,  'max_rb':25  \n",
    "\n",
    "In the scatter and boxplots 9 clusters fall outside the filtering parameters.  \n",
    "0 -> low genes  \n",
    "2 -> low genes, high Mt  \n",
    "9 -> low genes  \n",
    "10 -> high Rb, low genes  \n",
    "12 -> high Rb, low genes, low Mt  \n",
    "14 -> low Rb, low genes, high Mt  \n",
    "19 -> high Rb, low genes  \n",
    "20 -> low genes, high Mt  \n",
    "22 -> low Rb, low genes, low Mt  \n",
    "\n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME9_scatter_qc.png)  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME9_boxplot.png)  \n",
    "\n",
    "I want to check what they are before removing them.  \n",
    "\n",
    "0 -> Markers include mostly blood genes. Hbb-y etc. I'm not sure if it's blood or just ambient RNA, either way it is not important.  \n",
    "2 -> Markers are mitochondrial genes. Probably bad cells or just captured mitochondria.  \n",
    "9 -> Blood.  \n",
    "10 -> Low quality cells. No markers.  \n",
    "12 -> Low quality cells. No markers.  \n",
    "14 -> Low quality brain cells. Markers such as Fgf14. Groups with rest of Sox2+ cells. Very high Mt and low genes.  \n",
    "19 -> Low quality mesoderm/mesenchyme. Markers such as Twist1.  \n",
    "20 -> This cluster seems mixed. Around half is bad cells, the other half seem legitimate. Markers such as Isl1 and Ccdc141. Luckily this agrees with the box/scatterplots.  \n",
    "The bad cells fall outside of the cutoffs.  \n",
    "22 -> Blood.  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME9_umap.png)  \n",
    "\n",
    "I will mark the bad clusters as failed.:  \n",
    "['0','2','9','10','12','14','19','22']\n",
    "\n",
    "Cells after QC:  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME9_QC_umap.png)  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "Here cluster 16 stands out as being composed of likely doublets. In addition we can see small clumps of doublets around.  \n",
    "I filtered all cells by a doublet score of 0.18 and removed cluster 16.  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME9_pass_umap.png) ![](../markdown_images/mouse/mm10/markdown_plots/ME9_pass_doublet_umap.png)   \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME9_pass_doublet_boxplot.png)   \n",
    "\n",
    "## ME10\n",
    "At ME10 I chose the following filtering parameters:  \n",
    "'min_counts':5000, 'max_counts':100000, 'min_genes':2200, 'max_genes':10000,  'min_mito':.3,  'max_mito':2.2,   'min_rb':5,  'max_rb':28  \n",
    "\n",
    "In the scatter and boxplots 4 clusters fall outside the filtering parameters.  \n",
    "11 -> Low genes, high Mt  \n",
    "12 -> High Rb, low genes, low Mt  \n",
    "16 -> High Rb, low genes, low Mt  \n",
    "20 -> Low genes  \n",
    "\n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME10_scatter_qc.png)  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME10_boxplot.png)  \n",
    "\n",
    "Check them before removing.  \n",
    "\n",
    "11 -> Marker genes are all mitochondrial.  \n",
    "12 -> Low quality cells.  \n",
    "16 -> Low quality ectoderm. Epcam positive, but groups with low quality cells because of low genes, low Mt and low Rb.  \n",
    "20 -> Blood Hbb-bt.  \n",
    "\n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME10_umap.png)  \n",
    "\n",
    "I will mark the bad clusters as failed.:  \n",
    "['11','12','16','20']\n",
    "\n",
    "Cells after QC:  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME10_QC_umap.png)  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "Here cluster 12 stands out as being composed of likely doublets. In addition we can see small clumps of doublets around.  \n",
    "I filtered all cells by a doublet score of 0.18 and removed cluster 12.  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME10_pass_umap.png) ![](../markdown_images/mouse/mm10/markdown_plots/ME10_pass_doublet_umap.png)   \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME10_pass_doublet_boxplot.png)  \n",
    "\n",
    "## ME11\n",
    "At ME11 I chose the following filtering parameters:  \n",
    "'min_counts':6000, 'max_counts':100000, 'min_genes':3100, 'max_genes':10000,  'min_mito':.5,  'max_mito':2.5,   'min_rb':9,  'max_rb':25  \n",
    "\n",
    "In the scatter and boxplots 6 clusters fall outside the filtering parameters.  \n",
    "0 -> Low genes, low Mt  \n",
    "3 -> Low genes, high Mt  \n",
    "12 -> Low Rb    \n",
    "15 -> Low Genes, low Mt  \n",
    "16 -> High Rb, low genes, low Mt  \n",
    "20 -> Low genes, low Mt  \n",
    "\n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME11_scatter_qc.png)    \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME11_boxplot.png)  \n",
    "\n",
    "I want to check what they are before removing them.  \n",
    "0 -> High Hbb-y and other trash. Bad cells.  \n",
    "3 -> Mostly mito genes.  \n",
    "12 -> These cells seem good, they have markers. Their main difference is that they are predicted to be in G1 cell cycle phase. I will readjust the min_rb to 5 and keep them.  \n",
    "15 -> Blood.  \n",
    "16 -> Low quality ectoderm. Epcam positive.  \n",
    "20 -> Low quality cns. Tubb3\n",
    "\n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME11_umap.png)  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME11phase_umap.png)  \n",
    "\n",
    "Ajusted filtering parameters to:  \n",
    "'min_counts':6000, 'max_counts':100000, 'min_genes':3100, 'max_genes':10000,  'min_mito':.5,  'max_mito':2.5,   'min_rb':5,  'max_rb':25  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME11_scatter_adjusted_qc.png)   \n",
    "\n",
    "And I will mark the bad clusters as failed too.:  \n",
    "['0','3','15','16','20']\n",
    "\n",
    "Cells after QC:  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME11_QC_umap.png)  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "Here cluster 22 stands out as being composed of likely doublets. In addition we can see small clumps of doublets around.  \n",
    "I filtered all cells by a doublet score of 0.2 and removed cluster 22.  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME11_pass_umap.png) ![](../markdown_images/mouse/mm10/markdown_plots/ME11_pass_doublet_umap.png)   \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME11_pass_doublet_boxplot.png)  \n",
    " \n",
    "\n",
    "## ME12\n",
    "At ME12 I chose the following filtering parameters:  \n",
    "'min_counts':2000, 'max_counts':100000, 'min_genes':2200, 'max_genes':10000,  'min_mito':.5,  'max_mito':2.5,   'min_rb':9,  'max_rb':30  \n",
    "\n",
    "A lot of the clusters have a portion of cells with really high Mt. I don't believe these cells are good. They generally have low counts and low genes as well.  \n",
    "I checked if they were overrepresented for any cell cycle state, but they were not. I'm not exactly sure what to do with these cells. Traditionally I have kept  \n",
    "the filtering parameters fairly strict unless there was a clear reason not to do so. For now I'll follow that and assume that these cells are bad in some way.  \n",
    "\n",
    "As for the low Rb in cluster 11, this is mainly G1 cell cycle. So I will relax this filtering parameter, similarly to the previous stage. Set min_rb to 5.  \n",
    "\n",
    "0 -> Low genes, low Mt  \n",
    "1 -> Low genes, high Mt  \n",
    "11 -> Low Rb  \n",
    "15 -> Low Rb, low genes, low Mt  \n",
    "23 -> Low genes  \n",
    "24 -> Low Rb, low genes, low Mt  \n",
    "\n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME12_scatter_qc.png)  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME12_boxplot.png)  \n",
    "\n",
    "0 -> No real markers, just crap.  \n",
    "1 -> Mostly Mito genes.  \n",
    "11 -> G1 phase neurons/brain.  \n",
    "15 -> Blood.  \n",
    "23 -> Poor quality neuronal cells (Tubb3)\n",
    "24 -> Blood.  \n",
    "\n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME12_umap.png)  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME12phase_umap.png)  \n",
    "\n",
    "Ajusted filtering parameters to:  \n",
    "'min_counts':2000, 'max_counts':100000, 'min_genes':2200, 'max_genes':10000,  'min_mito':.5,  'max_mito':2.5,   'min_rb':5,  'max_rb':30   \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME12_scatter_adjusted_qc.png)  \n",
    "\n",
    "And I will mark the bad clusters as failed too.:  \n",
    "['0','1','15','23','24']\n",
    "\n",
    "Cells after QC:  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME12_QC_umap.png)  \n",
    "Just wanted to note the amount of failed cells here. These have low genes and a very high mitochondrial content.  \n",
    "This is generally believed to be indicative of bad/stressed cells. In addition this sample had way more cells than the other samples.  \n",
    "I would not be surprised if that contributed.  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "Here no cluster stands out as being composed of likely doublets. But we can see small clumps of doublets around.  \n",
    "I filtered all cells by a doublet score of 0.18.  \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME12_pass_umap.png) ![](../markdown_images/mouse/mm10/markdown_plots/ME12_pass_doublet_umap.png)   \n",
    "![](../markdown_images/mouse/mm10/markdown_plots/ME12_pass_doublet_boxplot.png) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2b11fab-44ce-4100-b7f4-aa28689364e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.getLogger('matplotlib.font_manager').setLevel(logging.ERROR)\n",
    "import scanpy as sc\n",
    "import scvelo as scv\n",
    "import scvi\n",
    "import seaborn as sns\n",
    "import plotly.express as px\n",
    "import numpy as np\n",
    "from dash import Dash, dcc, html, Input, Output\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "import gc\n",
    "os.environ['R_HOME'] = sys.exec_prefix+\"/lib/R/\"\n",
    "\n",
    "# Plotting\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "from matplotlib.colors import LinearSegmentedColormap, ListedColormap\n",
    "from matplotlib.lines import Line2D \n",
    "\n",
    "from copy import copy\n",
    "reds = copy(mpl.cm.Reds)\n",
    "reds.set_under(\"lightgray\")\n",
    "\n",
    "project_directory = '/Cranio_Lab/Louk_Seton/4_species_project'\n",
    "os.chdir(os.path.expanduser(\"~\")+project_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbb1cd63-6fbf-4587-890d-beb22fae5d94",
   "metadata": {},
   "outputs": [],
   "source": [
    "##TODO\n",
    "\n",
    "#try soft integration with just hvgs > look into better hvg selection\n",
    "#else hard integration with mira for example\n",
    "#try to automate plotting good DEGs\n",
    "#roughly annotate cells\n",
    "#plot annotated umap and dotplots\n",
    "\n",
    "#repeat on other organisms\n",
    "\n",
    "######Mouse mm10######\n",
    "start_time=time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\")\n",
    "print('start time:',start_time)\n",
    "\n",
    "sample_names = ['ME8','ME9','ME10','ME11','ME12'] #specify the sample names\n",
    "species = 'mouse' #specify the species\n",
    "genome = 'mm10' #specify the genome\n",
    "output_prefix = 'cellranger_related/cellranger_output/sc_rnaseq/' #specify the location of the cellranger output\n",
    "\n",
    "##plotting vars\n",
    "output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "genes_to_plot = ['Sox10','Plp1','Hbb-y','Epcam','Twist2','Prrx2','Tfap2b','Alx3','Dlx1','Dlx5','Meox2','Sox2',]\n",
    "dot_size = 1\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "##################################################################################################################\n",
    "cellranger_locs = []\n",
    "velocyto_locs = []\n",
    "\n",
    "for sample in sample_names: #simple loop to create a list with the location of the required files\n",
    "    cellranger_locs = cellranger_locs + [output_prefix+species+'/'+genome+'/'+sample+'/outs/filtered_feature_bc_matrix/']\n",
    "    velocyto_locs = velocyto_locs + [output_prefix+species+'/'+genome+'/'+sample+'/velocyto/'+sample+'.loom']\n",
    "\n",
    "adata_files={}\n",
    "print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Reading file(s)')\n",
    "for i,j,k in zip(sample_names, cellranger_locs, velocyto_locs):\n",
    "    ##Read and merge 10x and velocyto output, preserving gene names and ensemblid\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Sample ',i)\n",
    "    adata_cellr = sc.read_10x_mtx(j)\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Loaded 10x mtx')\n",
    "    adata_cellr.obs.index = adata_cellr.obs.index.str[:-2]\n",
    "    adata_cellr.var['gene_name'] = adata_cellr.var.index\n",
    "    adata_cellr.var.index = adata_cellr.var['gene_ids']\n",
    "    adata_veloc = sc.read(k)\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Loaded velocyto loom')\n",
    "    adata_veloc.obs.index = adata_veloc.obs.index.str.split(':').str[1]\n",
    "    adata_veloc.obs.index = adata_veloc.obs.index.str[:-1]\n",
    "    adata_veloc.var['gene_name'] = adata_veloc.var.index\n",
    "    adata_veloc.var.index = adata_veloc.var['Accession']\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Merging data')\n",
    "    adata_merge = scv.utils.merge(adata_cellr, adata_veloc)\n",
    "    adata_files['adata'+i]=adata_merge\n",
    "    adata_files['adata'+i].obs['sample']=i\n",
    "    adata_files['adata'+i].obs['ref_genome']=genome\n",
    "    adata_files['adata'+i].obs['barcode']=adata_files['adata'+i].obs.index\n",
    "    adata_files['adata'+i].obs.index=adata_files['adata'+i].obs['barcode']+'.'+adata_files['adata'+i].obs['sample']\n",
    "\n",
    "    adata_files['adata'+i].var.index = list(adata_files['adata'+i].var['gene_name'])#set gene_names as index\n",
    "\n",
    "    ##Calculate QC values\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Calculating QC values for cells')\n",
    "    adata_files['adata'+i].var['mt'] = adata_files['adata'+i].var['gene_name'].str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "    adata_files['adata'+i].var['rb'] = adata_files['adata'+i].var['gene_name'].str.startswith(('Rps','Rpl'))  # annotate the group of ribosomal proteins as 'rb'\n",
    "    adata_files['adata'+i].var['hb'] = adata_files['adata'+i].var['gene_name'].str.startswith(('Hbb','Hba')) #annotate\n",
    "    adata_files['adata'+i].var_names_make_unique()\n",
    "    adata_files['adata'+i].obs_names_make_unique()\n",
    "    sc.pp.calculate_qc_metrics(adata_files['adata'+i], qc_vars=[\"mt\", \"rb\", \"hb\"], inplace=True, log1p=True)\n",
    "    #clip to meaningful numbers, better for comparisons\n",
    "    adata_files['adata'+i].obs['pct_counts_mt'] = adata_files['adata'+i].obs['pct_counts_mt'].clip(upper = 20)\n",
    "    adata_files['adata'+i].obs['pct_counts_hb'] = adata_files['adata'+i].obs['pct_counts_hb'].clip(upper = 20)\n",
    "    adata_files['adata'+i].obs['pct_counts_rb'] = adata_files['adata'+i].obs['pct_counts_rb'].clip(upper = 40)\n",
    "    adata_files['adata'+i].obs['total_counts'] = adata_files['adata'+i].obs['total_counts'].clip(upper = 100000)\n",
    "\n",
    "    ##calculate doublet score\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Running scrublet for doublet detection')\n",
    "    sc.pp.scrublet(adata_files['adata'+i])\n",
    "\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Initial preprocess and clustering')\n",
    "    ##standard normalization\n",
    "    adata_files['adata'+i].layers[\"original_counts\"] = adata_files['adata'+i].X.copy() #store original counts\n",
    "    sc.pp.normalize_total(adata_files['adata'+i]) # Normalizing to median total counts\n",
    "    sc.pp.log1p(adata_files['adata'+i]) # Logarithmize the data\n",
    "    adata_files['adata'+i].layers[\"normalized_counts\"] = adata_files['adata'+i].X.copy()\n",
    "\n",
    "    ##highly variable genes\n",
    "    sc.pp.highly_variable_genes(adata_files['adata'+i], n_top_genes=2000,)\n",
    "\n",
    "    ##dimensionality reduction and clustering\n",
    "    sc.tl.pca(adata_files['adata'+i])\n",
    "    sc.pp.neighbors(adata_files['adata'+i])\n",
    "    sc.tl.umap(adata_files['adata'+i])\n",
    "    sc.tl.leiden(adata_files['adata'+i],)\n",
    "\n",
    "    ##cell cycle scoring\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Prepare cell cycle scoring')\n",
    "    cell_cycle_genes = [x.strip() for x in open('required_files/regev_lab_cell_cycle_genes.txt')]\n",
    "    s_genes = cell_cycle_genes[:43]\n",
    "    g2m_genes = cell_cycle_genes[43:]\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Cell cycle scoring')\n",
    "    sc.tl.score_genes_cell_cycle(adata_files['adata'+i], s_genes=s_genes, g2m_genes=g2m_genes)\n",
    "\n",
    "    \n",
    "    #plot pdfs of qc\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Plotting QC and saving to PDF')\n",
    "    with PdfPages(output_dir+i+'_QC.pdf') as pdf:\n",
    "        mpl.rcParams['axes.grid'] = False\n",
    "        ax = sc.pl.violin(adata_files['adata'+i],[\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\",\"pct_counts_hb\", \"pct_counts_rb\",'doublet_score'],\n",
    "                          stripplot=False, multi_panel=True,show=False,\n",
    "                          height=4, aspect=.8,col_wrap = 3)\n",
    "        plt.suptitle(i, x = 0.1)\n",
    "        pdf.savefig()  # saves the current figure into a pdf page\n",
    "        plt.close() # close the plot and page\n",
    "\n",
    "        ###Scatterplots\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap=scatter_cmap\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files['adata'+i].obs, x=x1,y=y1,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax1.set_ylabel(y1)\n",
    "        ax2.scatter(data=adata_files['adata'+i].obs, x=x2,y=y1,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax2.set_ylabel(y1)\n",
    "        ax3.scatter(data=adata_files['adata'+i].obs, x=x1,y=y2,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax3.set_xlabel(x1) \n",
    "        ax3.set_ylabel(y2)\n",
    "        ax4.scatter(data=adata_files['adata'+i].obs, x=x2,y=y2,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax4.set_xlabel(x2) \n",
    "        ax4.set_ylabel(y2)\n",
    "\n",
    "\n",
    "        fig.colorbar(mpl.cm.ScalarMappable(norm=mpl.colors.Normalize(0, max(adata_files['adata'+i].obs[c])), cmap=cmap),\n",
    "                     ax=ax1, orientation='vertical', label=c)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##nice jointplot shows good cells\n",
    "        mpl.rcParams['axes.grid'] = False\n",
    "        sns_plot = sns.jointplot(data=adata_files['adata'+i].obs,\n",
    "                                 x=\"log1p_total_counts\",\n",
    "                                 y=\"log1p_n_genes_by_counts\",\n",
    "                                 kind=\"hex\",\n",
    "                                 marginal_kws=dict(bins=100,),\n",
    "                                 joint_kws = dict(gridsize=50,),\n",
    "                                )\n",
    "        sns_plot.fig.suptitle(i,x=0.05)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##custom hvg plot\n",
    "        fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize = (8,4))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap = ListedColormap(['grey','black'])\n",
    "        c = 'highly_variable'\n",
    "        scatter = ax1.scatter(data=adata_files['adata'+i].var, x='means',y=\"dispersions\",\n",
    "                              c=c,s = s,cmap=cmap,linewidth=linewidth,rasterized=True)\n",
    "        ax1.set_xlabel('mean expression') \n",
    "        ax1.set_ylabel('dispersion') \n",
    "        ax1.legend(scatter.legend_elements()[0][0:2],['other genes','highly variable'],\n",
    "                   loc=\"upper right\",)\n",
    "        scatter = ax2.scatter(data=adata_files['adata'+i].var, x='means',y=\"dispersions_norm\",\n",
    "                              c=c,s = s,cmap=cmap,linewidth=linewidth,rasterized=True)\n",
    "        ax2.legend(scatter.legend_elements()[0][0:2],['other genes','highly variable'],\n",
    "                   loc=\"upper right\",)\n",
    "        ax2.set_xlabel('mean expression') \n",
    "        ax2.set_ylabel('dispersions_norm')\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##PCA variance plot\n",
    "        sc.pl.pca_variance_ratio(adata_files['adata'+i], log=True, show=False)\n",
    "        pdf.savefig()\n",
    "        plt.close()\n",
    "        \n",
    "        ##umaps\n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files['adata'+i],\n",
    "                        color=['doublet_score','leiden','total_counts','pct_counts_mt','pct_counts_rb','pct_counts_hb','phase',]+genes_to_plot,\n",
    "                        show = False,ncols = 2)\n",
    "        for p in ax:\n",
    "            p.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##Most highly expressed genes\n",
    "        plt.rcParams['figure.figsize'] = [8,4]\n",
    "        adata_files['adata'+i].X = adata_files['adata'+i].layers[\"original_counts\"].copy() #set raw counts to X because the layer argument is broken\n",
    "        ax = sc.pl.highest_expr_genes(adata_files['adata'+i], \n",
    "                                      gene_symbols = 'gene_name',\n",
    "                                      n_top=20,log=True,show=False)\n",
    "        ax.set_rasterized(True)\n",
    "        pdf.savefig()\n",
    "        plt.close()\n",
    "        ##return normalised counts to X\n",
    "        adata_files['adata'+i].X = adata_files['adata'+i].layers[\"normalized_counts\"].copy()\n",
    "\n",
    "        ##doublet scatterplots\n",
    "        ###Scatterplots\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap=scatter_cmap\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files['adata'+i].obs, x=x1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax2.scatter(data=adata_files['adata'+i].obs, x=x2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax3.scatter(data=adata_files['adata'+i].obs, x=y1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax3.set_xlabel(y1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax4.scatter(data=adata_files['adata'+i].obs, x=y2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax4.set_xlabel(y2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        fig.colorbar(mpl.cm.ScalarMappable(norm=mpl.colors.Normalize(0, max(adata_files['adata'+i].obs[c])), cmap=cmap),\n",
    "                     ax=ax1, orientation='vertical', label=c)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "        \n",
    "        #boxplot of doublet scores\n",
    "        ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"doublet_score\",\n",
    "            palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0)\n",
    "        cnum = 0\n",
    "        for line in ax.lines:\n",
    "            if line.get_linestyle() == 'None':\n",
    "                sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                              size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True)\n",
    "                cnum = cnum+1\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "    ##saving h5ad\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Saving data to h5ad file')\n",
    "    adata_files['adata'+i].write('h5ad_files/'+species+'/'+genome+'/'+i+'_before_filtering.h5ad')\n",
    "    \n",
    "\n",
    "    ##remove used variables from memory\n",
    "    del adata_cellr\n",
    "    del adata_veloc\n",
    "    del adata_merge\n",
    "    gc.collect()\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Loaded files for ',i)\n",
    "print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Finished Reading file(s)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e046a48f-a88b-4ad9-bb08-9af96dd9a452",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Filter using QC vars and make plots\n",
    "filter_vars = {'ME8':{'min_counts':10000, 'max_counts':100000, 'min_genes':2900, 'max_genes':10000,  'min_mito':.1,  'max_mito':1.5,   'min_rb':15,  'max_rb':31,\n",
    "                      'clusters':['12','15','18','20','22']},\n",
    "               'ME9':{'min_counts':9000, 'max_counts':100000, 'min_genes':4000, 'max_genes':10000,  'min_mito':.2,  'max_mito':2,   'min_rb':10,  'max_rb':25,\n",
    "                      'clusters':['0','2','9','10','12','14','19','22']}, \n",
    "               'ME10':{'min_counts':5000, 'max_counts':100000, 'min_genes':2200, 'max_genes':10000,  'min_mito':.3,  'max_mito':2.2,   'min_rb':5,  'max_rb':28,\n",
    "                      'clusters':['11','12','16','20']},\n",
    "               'ME11':{'min_counts':6000, 'max_counts':100000, 'min_genes':3100, 'max_genes':10000,  'min_mito':.5,  'max_mito':2.5,   'min_rb':5,  'max_rb':25,\n",
    "                      'clusters':['0','3','15','16','20']}, #adjusted\n",
    "               'ME12':{'min_counts':2000, 'max_counts':100000, 'min_genes':2200, 'max_genes':10000,  'min_mito':.5,  'max_mito':2.5,   'min_rb':5,  'max_rb':30,\n",
    "                      'clusters':['0','1','15','23','24']}, #adjusted\n",
    "}               \n",
    "filter_vars = pd.DataFrame(filter_vars).T\n",
    "\n",
    "\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "dot_size = 1\n",
    "linewidth = 0\n",
    "scatter_color = 'QC'\n",
    "output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "for sample in list(filter_vars.index):\n",
    "    min_counts = filter_vars.loc[sample,'min_counts']\n",
    "    max_counts = filter_vars.loc[sample,'max_counts']\n",
    "    min_genes = filter_vars.loc[sample,'min_genes']\n",
    "    max_genes = filter_vars.loc[sample,'max_genes']\n",
    "    min_mito = filter_vars.loc[sample,'min_mito']\n",
    "    max_mito = filter_vars.loc[sample,'max_mito']\n",
    "    min_rb = filter_vars.loc[sample,'min_rb']\n",
    "    max_rb = filter_vars.loc[sample,'max_rb']\n",
    "\n",
    "    conditions = [\n",
    "    (adata_files['adata'+sample].obs['predicted_doublet'] == True),\n",
    "    (adata_files['adata'+sample].obs['n_genes_by_counts'] < min_genes),\n",
    "    (adata_files['adata'+sample].obs['n_genes_by_counts'] > max_genes),\n",
    "    (adata_files['adata'+sample].obs['total_counts'] < min_counts),\n",
    "    (adata_files['adata'+sample].obs['total_counts'] > max_counts),\n",
    "    (adata_files['adata'+sample].obs['pct_counts_mt'] > max_mito),\n",
    "    (adata_files['adata'+sample].obs['pct_counts_mt'] < min_mito),\n",
    "    (adata_files['adata'+sample].obs['pct_counts_rb'] < min_rb),\n",
    "    (adata_files['adata'+sample].obs['pct_counts_rb'] > max_rb),\n",
    "    (adata_files['adata'+sample].obs['pct_counts_mt'] <= max_mito) & (adata_files['adata'+sample].obs['pct_counts_mt'] >= min_mito) & (adata_files['adata'+sample].obs['n_genes_by_counts'] >= min_genes) & (adata_files['adata'+sample].obs['n_genes_by_counts'] <= max_genes) & \n",
    "    (adata_files['adata'+sample].obs['total_counts'] >= min_counts) & (adata_files['adata'+sample].obs['total_counts'] <= max_counts) & (adata_files['adata'+sample].obs['pct_counts_rb'] >= min_rb) & (adata_files['adata'+sample].obs['pct_counts_rb'] <= max_rb) & \n",
    "    (adata_files['adata'+sample].obs['predicted_doublet'] != True)\n",
    "    ]\n",
    "    \n",
    "    values = ['Doublet','Low_nFeature', 'High_nFeature', 'Low_Counts','High_Counts', 'High_MT','Low_MT', 'Low_RB' ,'High_RB' ,'Pass']\n",
    "    adata_files['adata'+sample].obs['QC'] = np.select(conditions, values)\n",
    "    adata_files['adata'+sample].obs.loc[adata_files['adata'+sample].obs['leiden'].isin(filter_vars.loc[sample,'clusters']), 'QC'] = \"Bad_Cluster\"\n",
    "    adata_files['adata'+sample].obs['QC'] = adata_files['adata'+sample].obs['QC'].astype('category')\n",
    "\n",
    "    colors = {'Doublet':'#bc80bd', \n",
    "                  'Low_nFeature':'#d9d9d9', \n",
    "                  'High_nFeature':'#8dd3c7', \n",
    "                  'Low_Counts':'#ffffb3',\n",
    "                  'High_Counts':'#bebada',\n",
    "                  'High_MT':'#fb8072',\n",
    "                  'Low_MT':'#80b1d3',\n",
    "                  'Low_RB':'#fdb462',\n",
    "                  'High_RB':'#fccde5',\n",
    "                  'Bad_Cluster':'#000000',\n",
    "                  'Pass':'#b3de69',\n",
    "                 }\n",
    "    with PdfPages(output_dir+sample+'_filter_params.pdf') as pdf:        \n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files['adata'+sample], color = ['QC'],palette = colors, show=False)\n",
    "        ax.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150,bbox_inches='tight')  # saves the current figure into a pdf page\n",
    "        plt.close() # close the plot and page\n",
    "        \n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files['adata'+sample], color = ['n_genes_by_counts','doublet_score','phase','pct_counts_hb'],ncols=2,show=False)\n",
    "        for p in ax:\n",
    "            p.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150)  # saves the current figure into a pdf page\n",
    "        plt.close() # close the plot and page\n",
    "\n",
    "        ###Scatterplots\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files['adata'+sample].obs, x=x1,y=y1,\n",
    "                    c = adata_files['adata'+sample].obs[c].map(colors),s = s,linewidth=linewidth,rasterized=True)\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax1.set_ylabel(y1)\n",
    "        ax2.scatter(data=adata_files['adata'+sample].obs, x=x2,y=y1,\n",
    "                    c = adata_files['adata'+sample].obs[c].map(colors),s = s,linewidth=linewidth,rasterized=True)\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax2.set_ylabel(y1)\n",
    "        ax3.scatter(data=adata_files['adata'+sample].obs, x=x1,y=y2,\n",
    "                    c = adata_files['adata'+sample].obs[c].map(colors),s = s,linewidth=linewidth,rasterized=True)\n",
    "        ax3.set_xlabel(x1) \n",
    "        ax3.set_ylabel(y2)\n",
    "        scatter = ax4.scatter(data=adata_files['adata'+sample].obs, x=x2,y=y2,\n",
    "                    c = adata_files['adata'+sample].obs[c].map(colors),s = s,linewidth=linewidth,rasterized=True)\n",
    "        ax4.set_xlabel(x2) \n",
    "        ax4.set_ylabel(y2)\n",
    "        handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=8) for k, v in colors.items()]\n",
    "        ax4.legend(title='color', handles=handles, bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6840b256-d26b-4348-a4de-c783458d6ec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "##doublet stuff and some more plots\n",
    "genes_to_plot = ['Sox10','Plp1','Hbb-y','Epcam','Twist2','Prrx2','Tfap2b','Alx3','Dlx1','Dlx5','Meox2','Sox2',]\n",
    "dot_size = 1\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "\n",
    "for i in sample_names:\n",
    "    sn = 'adata'+i #I got tired of long ass code so shortening the samplename\n",
    "    adata_files[sn+'_pass'] = adata_files[sn][adata_files[sn].obs['QC']=='Pass'].copy() #filter cells\n",
    "    sn = sn+'_pass'\n",
    "    adata_files[sn].X = adata_files[sn].layers[\"original_counts\"].copy() #restore original counts\n",
    "    sc.pp.scrublet(adata_files[sn])\n",
    "\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Initial preprocess and clustering')\n",
    "    ##standard normalization\n",
    "    sc.pp.normalize_total(adata_files[sn]) # Normalizing to median total counts\n",
    "    sc.pp.log1p(adata_files[sn]) # Logarithmize the data\n",
    "    adata_files[sn].layers[\"normalized_counts\"] = adata_files[sn].X.copy()\n",
    "\n",
    "    ##highly variable genes\n",
    "    sc.pp.highly_variable_genes(adata_files[sn], n_top_genes=2000,)\n",
    "\n",
    "    ##dimensionality reduction and clustering\n",
    "    sc.tl.pca(adata_files[sn])\n",
    "    sc.pp.neighbors(adata_files[sn])\n",
    "    sc.tl.umap(adata_files[sn])\n",
    "    sc.tl.leiden(adata_files[sn],resolution = 1, key_added = 'leiden_filt')\n",
    "    sc.tl.leiden(adata_files[sn],resolution = 2, key_added = 'leiden_filt_highres')\n",
    "\n",
    "    ##plotting new stuff\n",
    "    output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "    !mkdir -p {output_dir}\n",
    "\n",
    "    with PdfPages(output_dir+i+'_doublet_check.pdf') as pdf:\n",
    "        ##umaps\n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files[sn],\n",
    "                        color=['doublet_score','leiden','total_counts','leiden_filt','pct_counts_mt','leiden_filt_highres','pct_counts_rb','pct_counts_hb','phase',]+genes_to_plot,\n",
    "                        show = False,ncols = 2)\n",
    "        for p in ax:\n",
    "            p.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "        \n",
    "        #boxplot of doublet scores\n",
    "        plt.rcParams['figure.figsize'] = [8,4]\n",
    "        ax = sns.boxplot(data=adata_files[sn].obs, x=\"leiden\", y=\"doublet_score\",\n",
    "            palette = adata_files[sn].uns['leiden_colors'], \n",
    "                 fliersize=0)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ax = sns.boxplot(data=adata_files[sn].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "            palette = adata_files[sn].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "        cnum = 0\n",
    "        for line in ax.lines:\n",
    "            if line.get_linestyle() == 'None':\n",
    "                sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                              size = 1.5, color = adata_files[sn].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "                cnum = cnum+1\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ax = sns.boxplot(data=adata_files[sn].obs, x=\"leiden_filt_highres\", y=\"doublet_score\",\n",
    "            palette = adata_files[sn].uns['leiden_filt_highres_colors'], \n",
    "                 fliersize=0)\n",
    "        cnum = 0\n",
    "        for line in ax.lines:\n",
    "            if line.get_linestyle() == 'None':\n",
    "                sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                              size = 1.5, color = adata_files[sn].uns['leiden_filt_highres_colors'][cnum], rasterized = True)\n",
    "                cnum = cnum+1\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##doublet scatterplots\n",
    "        ###Scatterplots\n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap=scatter_cmap\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files[sn].obs, x=x1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax2.scatter(data=adata_files[sn].obs, x=x2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax3.scatter(data=adata_files[sn].obs, x=y1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax3.set_xlabel(y1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax4.scatter(data=adata_files[sn].obs, x=y2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax4.set_xlabel(y2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        fig.colorbar(mpl.cm.ScalarMappable(norm=mpl.colors.Normalize(0, max(adata_files[sn].obs[c])), cmap=cmap),\n",
    "                     ax=ax1, orientation='vertical', label=c)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b99c5658-1a44-4a22-bebb-a497830a5c77",
   "metadata": {},
   "outputs": [],
   "source": [
    "doublet_vars = {'ME8_pass':{'doublet_cutoff':.15, 'doublet_clusters':['14']},\n",
    "                'ME9_pass':{'doublet_cutoff':.18, 'doublet_clusters':['16']},\n",
    "                'ME10_pass':{'doublet_cutoff':.18, 'doublet_clusters':['12']},\n",
    "                'ME11_pass':{'doublet_cutoff':.2, 'doublet_clusters':['22']},\n",
    "                'ME12_pass':{'doublet_cutoff':.18, 'doublet_clusters':None},\n",
    "} \n",
    "doublet_vars = pd.DataFrame(doublet_vars).T\n",
    "\n",
    "genes_to_plot = ['Sox10','Plp1','Epcam','Twist2','Prrx2','Tfap2b','Alx3','Dlx1','Dlx5','Meox2','Sox2',]\n",
    "\n",
    "for index in list(doublet_vars.index):\n",
    "    adata_obj = 'adata'+index\n",
    "    if 'pass' in adata_obj:\n",
    "        doublet_cutoff = doublet_vars.loc[index,'doublet_cutoff']\n",
    "        doublet_clusters = doublet_vars.loc[index,'doublet_clusters']\n",
    "        adata_files[adata_obj].obs['doublet_cutoff'] = 'pass'\n",
    "        adata_files[adata_obj].obs.loc[adata_files[adata_obj].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "        if isinstance(doublet_clusters, list):\n",
    "            adata_files[adata_obj].obs.loc[adata_files[adata_obj].obs['leiden_filt'].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "        adata_files[adata_obj].obs['doublet_cutoff'] = adata_files[adata_obj].obs['doublet_cutoff'].astype('category')\n",
    "\n",
    "        sn = adata_obj+'_doublet'\n",
    "        adata_files[sn] = adata_files[adata_obj][adata_files[adata_obj].obs['doublet_cutoff'] == 'pass'].copy()\n",
    "        adata_files[sn].X = adata_files[sn].layers[\"original_counts\"].copy() #restore original counts\n",
    "        print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Initial preprocess and clustering')\n",
    "        ##standard normalization\n",
    "        sc.pp.normalize_total(adata_files[sn]) # Normalizing to median total counts\n",
    "        sc.pp.log1p(adata_files[sn]) # Logarithmize the data\n",
    "        adata_files[sn].layers[\"normalized_counts\"] = adata_files[sn].X.copy()\n",
    "\n",
    "        ##highly variable genes\n",
    "        sc.pp.highly_variable_genes(adata_files[sn], n_top_genes=1000,)\n",
    "\n",
    "        ##dimensionality reduction and clustering\n",
    "        sc.tl.pca(adata_files[sn])\n",
    "        sc.pp.neighbors(adata_files[sn])\n",
    "        sc.tl.umap(adata_files[sn])\n",
    "        sc.tl.leiden(adata_files[sn],resolution = 1, key_added = 'leiden_post_QC')\n",
    "        sc.tl.leiden(adata_files[sn],resolution = 2, key_added = 'leiden_post_QC_highres')\n",
    "\n",
    "        ##plotting new stuff\n",
    "        output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "        !mkdir -p {output_dir}\n",
    "\n",
    "        i = adata_obj.split('adata')[1].split('_')[0]\n",
    "        with PdfPages(output_dir+i+'_post_QC.pdf') as pdf:\n",
    "            ##umaps\n",
    "            plt.rcParams['figure.figsize'] = [5,4]\n",
    "            ax = sc.pl.umap(adata_files[sn],\n",
    "                            color=['leiden_post_QC','leiden_post_QC_highres','phase',]+genes_to_plot,\n",
    "                            show = False,ncols = 2)\n",
    "            for p in ax:\n",
    "                p.set_rasterized(True)\n",
    "            pdf.savefig(dpi=150)\n",
    "            plt.close()\n",
    "        \n",
    "        adata_files[sn].write('h5ad_files/'+species+'/'+genome+'/'+i+'_after_filtering.h5ad')\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b6899a2-33c4-4ad7-98d2-51b0f1fa7388",
   "metadata": {},
   "outputs": [],
   "source": [
    "## code for any markdown figures ##\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "## let's start with ME8\n",
    "i = 'ME8'\n",
    "filter_vars = {'ME8':{'min_counts':10000, 'max_counts':100000, 'min_genes':2900, 'max_genes':10000,  'min_mito':.1,  'max_mito':1.5,   'min_rb':15,  'max_rb':31}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files['adata'+i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files['adata'+i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files['adata'+i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,3)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files['adata'+i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplot of doublet scores\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = ['leiden','Epcam','Pou3f1','Myl4','pct_counts_hb'],groups = ['12','15','18','20','22'], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "i = 'ME9'\n",
    "filter_vars = {'ME9':{'min_counts':9000, 'max_counts':100000, 'min_genes':4000, 'max_genes':10000,  'min_mito':.2,  'max_mito':2,   'min_rb':10,  'max_rb':25}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files['adata'+i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files['adata'+i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files['adata'+i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,3)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files['adata'+i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplot of doublet scores\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = ['leiden','Hbb-y','Fgf14','Twist1','Isl1'],groups = ['0','2','9','10','12','14','19','20','22'], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "i = 'ME10'\n",
    "filter_vars = {'ME10':{'min_counts':5000, 'max_counts':100000, 'min_genes':2200, 'max_genes':10000,  'min_mito':.3,  'max_mito':2.2,   'min_rb':5,  'max_rb':28}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files['adata'+i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files['adata'+i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files['adata'+i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,3)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files['adata'+i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplot of doublet scores\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = ['leiden','Epcam','Hbb-bt'],groups = ['11','12','16','20',], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "i = 'ME11'\n",
    "filter_vars = {'ME11':{'min_counts':6000, 'max_counts':100000, 'min_genes':3100, 'max_genes':10000,  'min_mito':.5,  'max_mito':2.5,   'min_rb':9,  'max_rb':25}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files['adata'+i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files['adata'+i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files['adata'+i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,3)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files['adata'+i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplot of doublet scores\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = ['leiden','Hbb-y','Epcam','pct_counts_hb'],groups = ['0','3','12','15','16','20'], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = ['phase',], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'phase_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "filter_vars = {'ME11':{'min_counts':6000, 'max_counts':100000, 'min_genes':3100, 'max_genes':10000,  'min_mito':.5,  'max_mito':2.5,   'min_rb':5,  'max_rb':25}, #adjusted\n",
    "}   \n",
    "\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files['adata'+i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files['adata'+i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files['adata'+i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,3)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files['adata'+i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_adjusted_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "i = 'ME12'\n",
    "filter_vars = {'ME12':{'min_counts':2000, 'max_counts':100000, 'min_genes':2200, 'max_genes':10000,  'min_mito':.5,  'max_mito':2.5,   'min_rb':9,  'max_rb':30}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files['adata'+i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files['adata'+i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files['adata'+i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,15)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files['adata'+i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplot of doublet scores\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = ['leiden','Tubb3','pct_counts_hb'],groups = ['0','1','11','15','23','24'], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = ['phase',], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'phase_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "filter_vars = {'ME12':{'min_counts':2000, 'max_counts':100000, 'min_genes':2200, 'max_genes':10000,  'min_mito':.5,  'max_mito':2.5,   'min_rb':5,  'max_rb':30}, #adjusted\n",
    "}   \n",
    " \n",
    "\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files['adata'+i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files['adata'+i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files['adata'+i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,15)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files['adata'+i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files['adata'+i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_adjusted_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "##Post filter plots\n",
    "filter_vars = {'ME8':{'min_counts':10000, 'max_counts':100000, 'min_genes':2900, 'max_genes':10000,  'min_mito':.1,  'max_mito':1.5,   'min_rb':15,  'max_rb':31,\n",
    "                      'clusters':['12','15','18','20','22']},\n",
    "               'ME9':{'min_counts':9000, 'max_counts':100000, 'min_genes':4000, 'max_genes':10000,  'min_mito':.2,  'max_mito':2,   'min_rb':10,  'max_rb':25,\n",
    "                      'clusters':['0','2','9','10','12','14','19','22']}, \n",
    "               'ME10':{'min_counts':5000, 'max_counts':100000, 'min_genes':2200, 'max_genes':10000,  'min_mito':.3,  'max_mito':2.2,   'min_rb':5,  'max_rb':28,\n",
    "                      'clusters':['11','12','16','20']},\n",
    "               'ME11':{'min_counts':6000, 'max_counts':100000, 'min_genes':3100, 'max_genes':10000,  'min_mito':.5,  'max_mito':2.5,   'min_rb':5,  'max_rb':25,\n",
    "                      'clusters':['0','3','15','16','20']}, #adjusted\n",
    "               'ME12':{'min_counts':2000, 'max_counts':100000, 'min_genes':2200, 'max_genes':10000,  'min_mito':.5,  'max_mito':2.5,   'min_rb':5,  'max_rb':30,\n",
    "                      'clusters':['0','1','15','23','24']}, #adjusted\n",
    "}\n",
    "\n",
    "filter_vars = pd.DataFrame(filter_vars).T\n",
    "\n",
    "for sample in list(filter_vars.index):\n",
    "    min_counts = filter_vars.loc[sample,'min_counts']\n",
    "    max_counts = filter_vars.loc[sample,'max_counts']\n",
    "    min_genes = filter_vars.loc[sample,'min_genes']\n",
    "    max_genes = filter_vars.loc[sample,'max_genes']\n",
    "    min_mito = filter_vars.loc[sample,'min_mito']\n",
    "    max_mito = filter_vars.loc[sample,'max_mito']\n",
    "    min_rb = filter_vars.loc[sample,'min_rb']\n",
    "    max_rb = filter_vars.loc[sample,'max_rb']\n",
    "\n",
    "    conditions = [\n",
    "    (adata_files['adata'+sample].obs['predicted_doublet'] == True),\n",
    "    (adata_files['adata'+sample].obs['n_genes_by_counts'] < min_genes),\n",
    "    (adata_files['adata'+sample].obs['n_genes_by_counts'] > max_genes),\n",
    "    (adata_files['adata'+sample].obs['total_counts'] < min_counts),\n",
    "    (adata_files['adata'+sample].obs['total_counts'] > max_counts),\n",
    "    (adata_files['adata'+sample].obs['pct_counts_mt'] > max_mito),\n",
    "    (adata_files['adata'+sample].obs['pct_counts_mt'] < min_mito),\n",
    "    (adata_files['adata'+sample].obs['pct_counts_rb'] < min_rb),\n",
    "    (adata_files['adata'+sample].obs['pct_counts_rb'] > max_rb),\n",
    "    (adata_files['adata'+sample].obs['pct_counts_mt'] <= max_mito) & (adata_files['adata'+sample].obs['pct_counts_mt'] >= min_mito) & (adata_files['adata'+sample].obs['n_genes_by_counts'] >= min_genes) & (adata_files['adata'+sample].obs['n_genes_by_counts'] <= max_genes) & \n",
    "    (adata_files['adata'+sample].obs['total_counts'] >= min_counts) & (adata_files['adata'+sample].obs['total_counts'] <= max_counts) & (adata_files['adata'+sample].obs['pct_counts_rb'] >= min_rb) & (adata_files['adata'+sample].obs['pct_counts_rb'] <= max_rb) & \n",
    "    (adata_files['adata'+sample].obs['predicted_doublet'] != True)\n",
    "    ]\n",
    "    \n",
    "    values = ['Fail','Fail', 'Fail', 'Fail','Fail', 'Fail','Fail', 'Fail' ,'Fail' ,'Pass']\n",
    "    adata_files['adata'+sample].obs['QC'] = np.select(conditions, values)\n",
    "    adata_files['adata'+sample].obs.loc[adata_files['adata'+sample].obs['leiden'].isin(filter_vars.loc[sample,'clusters']), 'QC'] = \"Fail\"\n",
    "    adata_files['adata'+sample].obs['QC'] = adata_files['adata'+sample].obs['QC'].astype('category')\n",
    "\n",
    "    plt.rcParams['figure.figsize'] = [4,3]\n",
    "    ax = sc.pl.umap(adata_files['adata'+sample], color = ['QC',], ncols = 3,show = False)\n",
    "    plt.savefig(output_dir+sample+'_QC_umap.png', dpi = 80,bbox_inches='tight')\n",
    "    plt.close()\n",
    "\n",
    "\n",
    "####Doublet stuff####\n",
    "#ME8\n",
    "i = 'ME8_pass'\n",
    "doublet_cutoff = .15\n",
    "doublet_clusters = ['14']\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = ['leiden_filt','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = 0.05)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files['adata'+i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files['adata'+i].obs.loc[adata_files['adata'+i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files['adata'+i].obs.loc[adata_files['adata'+i].obs['leiden_filt'].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files['adata'+i].obs['doublet_cutoff'] = adata_files['adata'+i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "\n",
    "#ME9\n",
    "i = 'ME9_pass'\n",
    "doublet_cutoff = .18\n",
    "doublet_clusters = ['16']\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = ['leiden_filt','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = 0.05)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files['adata'+i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files['adata'+i].obs.loc[adata_files['adata'+i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files['adata'+i].obs.loc[adata_files['adata'+i].obs['leiden_filt'].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files['adata'+i].obs['doublet_cutoff'] = adata_files['adata'+i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "#ME10\n",
    "i = 'ME10_pass'\n",
    "doublet_cutoff = .18\n",
    "doublet_clusters = ['12']\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = ['leiden_filt','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = 0.05)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files['adata'+i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files['adata'+i].obs.loc[adata_files['adata'+i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files['adata'+i].obs.loc[adata_files['adata'+i].obs['leiden_filt'].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files['adata'+i].obs['doublet_cutoff'] = adata_files['adata'+i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "#ME11\n",
    "i = 'ME11_pass'\n",
    "doublet_cutoff = .2\n",
    "doublet_clusters = ['22']\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = ['leiden_filt','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = 0.05)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files['adata'+i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files['adata'+i].obs.loc[adata_files['adata'+i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files['adata'+i].obs.loc[adata_files['adata'+i].obs['leiden_filt'].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files['adata'+i].obs['doublet_cutoff'] = adata_files['adata'+i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "#ME12\n",
    "i = 'ME12_pass'\n",
    "doublet_cutoff = .18\n",
    "#doublet_clusters = ['22']\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = ['leiden_filt','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = 0.05)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files['adata'+i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files['adata'+i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files['adata'+i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files['adata'+i].obs.loc[adata_files['adata'+i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "#adata_files['adata'+i].obs.loc[adata_files['adata'+i].obs['leiden_filt'].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files['adata'+i].obs['doublet_cutoff'] = adata_files['adata'+i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files['adata'+i], color = 'doublet_cutoff', size = 4, show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e99e66fc-1f2b-4118-867d-36e6453b4e74",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f14cd6d-8f6d-4fe2-9d26-5acbe851924f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c882e92b-2a43-4f10-b0e2-48e0c35387df",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43e0f195-d760-4d35-81e5-8991510ec2d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db8367b0-7bc5-4e37-8cac-0818558e2aef",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1277bbb-7bb6-4a90-8781-d55a1ddc51e7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9eea6d6d-4857-44e6-a83b-50c1b4ed970d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13e65da9-2ed5-4100-9414-ce520f05f6d3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0336dadc-cf27-45a1-a0e5-807ff00c30d8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9145b80-8993-4a74-b0db-5786589f629e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "686e57a3-7696-4a4e-af06-bbdfc08aeb71",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dadfb49-2b44-426b-82b0-452a609cd587",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "421fbac5-8ddf-461b-b556-ef7311dbca25",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15c33e3e-deb3-4642-9a46-f762db5c7a3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "######Mouse mm39######\n",
    "start_time=time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\")\n",
    "print('start time:',start_time)\n",
    "\n",
    "sample_names = ['ME8','ME9','ME10','ME11','ME12'] #specify the sample names\n",
    "species = 'mouse' #specify the species\n",
    "genome = 'mm39' #specify the genome\n",
    "output_prefix = 'cellranger_related/cellranger_output/sc_rnaseq/' #specify the location of the cellranger output\n",
    "\n",
    "\n",
    "##plotting vars\n",
    "output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "genes_to_plot = ['Sox10','Plp1','Hbb-y','Epcam','Twist2','Prrx2','Tfap2b','Alx3','Dlx1','Dlx5','Meox2','Sox2',]\n",
    "dot_size = 1\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "##################################################################################################################\n",
    "cellranger_locs = []\n",
    "velocyto_locs = []\n",
    "\n",
    "for sample in sample_names: #simple loop to create a list with the location of the required files\n",
    "    cellranger_locs = cellranger_locs + [output_prefix+species+'/'+genome+'/'+sample+'/outs/filtered_feature_bc_matrix/']\n",
    "    velocyto_locs = velocyto_locs + [output_prefix+species+'/'+genome+'/'+sample+'/velocyto/'+sample+'.loom']\n",
    "\n",
    "adata_files={}\n",
    "print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Reading file(s)')\n",
    "for i,j,k in zip(sample_names, cellranger_locs, velocyto_locs):\n",
    "    ##Read and merge 10x and velocyto output, preserving gene names and ensemblid\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Sample ',i)\n",
    "    adata_cellr = sc.read_10x_mtx(j)\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Loaded 10x mtx')\n",
    "    adata_cellr.obs.index = adata_cellr.obs.index.str[:-2]\n",
    "    adata_cellr.var['gene_name'] = adata_cellr.var.index\n",
    "    adata_cellr.var.index = adata_cellr.var['gene_ids']\n",
    "    adata_veloc = sc.read(k)\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Loaded velocyto loom')\n",
    "    adata_veloc.obs.index = adata_veloc.obs.index.str.split(':').str[1]\n",
    "    adata_veloc.obs.index = adata_veloc.obs.index.str[:-1]\n",
    "    adata_veloc.var['gene_name'] = adata_veloc.var.index\n",
    "    adata_veloc.var.index = adata_veloc.var['Accession']\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Merging data')\n",
    "    adata_merge = scv.utils.merge(adata_cellr, adata_veloc)\n",
    "    adata_files['adata'+i]=adata_merge\n",
    "    adata_files['adata'+i].obs['sample']=i\n",
    "    adata_files['adata'+i].obs['ref_genome']=genome\n",
    "    adata_files['adata'+i].obs['barcode']=adata_files['adata'+i].obs.index\n",
    "    adata_files['adata'+i].obs.index=adata_files['adata'+i].obs['barcode']+'.'+adata_files['adata'+i].obs['sample']\n",
    "\n",
    "    adata_files['adata'+i].var.index = list(adata_files['adata'+i].var['gene_name'])#set gene_names as index\n",
    "\n",
    "    ##Calculate QC values\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Calculating QC values for cells')\n",
    "    adata_files['adata'+i].var['mt'] = adata_files['adata'+i].var['gene_name'].str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "    adata_files['adata'+i].var['rb'] = adata_files['adata'+i].var['gene_name'].str.startswith(('Rps','Rpl'))  # annotate the group of ribosomal proteins as 'rb'\n",
    "    adata_files['adata'+i].var['hb'] = adata_files['adata'+i].var['gene_name'].str.startswith(('Hbb','Hba')) #annotate\n",
    "    adata_files['adata'+i].var_names_make_unique()\n",
    "    adata_files['adata'+i].obs_names_make_unique()\n",
    "    sc.pp.calculate_qc_metrics(adata_files['adata'+i], qc_vars=[\"mt\", \"rb\", \"hb\"], inplace=True, log1p=True)\n",
    "    #clip to meaningful numbers, better for comparisons\n",
    "    adata_files['adata'+i].obs['pct_counts_mt'] = adata_files['adata'+i].obs['pct_counts_mt'].clip(upper = 20)\n",
    "    adata_files['adata'+i].obs['pct_counts_hb'] = adata_files['adata'+i].obs['pct_counts_hb'].clip(upper = 20)\n",
    "    adata_files['adata'+i].obs['pct_counts_rb'] = adata_files['adata'+i].obs['pct_counts_rb'].clip(upper = 40)\n",
    "    adata_files['adata'+i].obs['total_counts'] = adata_files['adata'+i].obs['total_counts'].clip(upper = 100000)\n",
    "\n",
    "    ##calculate doublet score\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Running scrublet for doublet detection')\n",
    "    sc.pp.scrublet(adata_files['adata'+i])\n",
    "\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Initial preprocess and clustering')\n",
    "    ##standard normalization\n",
    "    adata_files['adata'+i].layers[\"original_counts\"] = adata_files['adata'+i].X.copy() #store original counts\n",
    "    sc.pp.normalize_total(adata_files['adata'+i]) # Normalizing to median total counts\n",
    "    sc.pp.log1p(adata_files['adata'+i]) # Logarithmize the data\n",
    "    adata_files['adata'+i].layers[\"normalized_counts\"] = adata_files['adata'+i].X.copy()\n",
    "\n",
    "    ##highly variable genes\n",
    "    sc.pp.highly_variable_genes(adata_files['adata'+i], n_top_genes=2000,)\n",
    "\n",
    "    ##dimensionality reduction and clustering\n",
    "    sc.tl.pca(adata_files['adata'+i])\n",
    "    sc.pp.neighbors(adata_files['adata'+i])\n",
    "    sc.tl.umap(adata_files['adata'+i])\n",
    "    sc.tl.leiden(adata_files['adata'+i],)\n",
    "\n",
    "    ##cell cycle scoring\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Prepare cell cycle scoring')\n",
    "    cell_cycle_genes = [x.strip() for x in open('required_files/regev_lab_cell_cycle_genes.txt')]\n",
    "    s_genes = cell_cycle_genes[:43]\n",
    "    g2m_genes = cell_cycle_genes[43:]\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Cell cycle scoring')\n",
    "    sc.tl.score_genes_cell_cycle(adata_files['adata'+i], s_genes=s_genes, g2m_genes=g2m_genes)\n",
    "\n",
    "    \n",
    "    #plot pdfs of qc\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Plotting QC and saving to PDF')\n",
    "    with PdfPages(output_dir+i+'_QC.pdf') as pdf:\n",
    "        mpl.rcParams['axes.grid'] = False\n",
    "        ax = sc.pl.violin(adata_files['adata'+i],[\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\",\"pct_counts_hb\", \"pct_counts_rb\",'doublet_score'],\n",
    "                          stripplot=False, multi_panel=True,show=False,\n",
    "                          height=4, aspect=.8,col_wrap = 3)\n",
    "        plt.suptitle(i, x = 0.1)\n",
    "        pdf.savefig()  # saves the current figure into a pdf page\n",
    "        plt.close() # close the plot and page\n",
    "\n",
    "        ###Scatterplots\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap=scatter_cmap\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files['adata'+i].obs, x=x1,y=y1,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax1.set_ylabel(y1)\n",
    "        ax2.scatter(data=adata_files['adata'+i].obs, x=x2,y=y1,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax2.set_ylabel(y1)\n",
    "        ax3.scatter(data=adata_files['adata'+i].obs, x=x1,y=y2,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax3.set_xlabel(x1) \n",
    "        ax3.set_ylabel(y2)\n",
    "        ax4.scatter(data=adata_files['adata'+i].obs, x=x2,y=y2,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax4.set_xlabel(x2) \n",
    "        ax4.set_ylabel(y2)\n",
    "\n",
    "\n",
    "        fig.colorbar(mpl.cm.ScalarMappable(norm=mpl.colors.Normalize(0, max(adata_files['adata'+i].obs[c])), cmap=cmap),\n",
    "                     ax=ax1, orientation='vertical', label=c)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##nice jointplot shows good cells\n",
    "        mpl.rcParams['axes.grid'] = False\n",
    "        sns_plot = sns.jointplot(data=adata_files['adata'+i].obs,\n",
    "                                 x=\"log1p_total_counts\",\n",
    "                                 y=\"log1p_n_genes_by_counts\",\n",
    "                                 kind=\"hex\",\n",
    "                                 marginal_kws=dict(bins=100,),\n",
    "                                 joint_kws = dict(gridsize=50,),\n",
    "                                )\n",
    "        sns_plot.fig.suptitle(i,x=0.05)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##custom hvg plot\n",
    "        fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize = (8,4))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap = ListedColormap(['grey','black'])\n",
    "        c = 'highly_variable'\n",
    "        scatter = ax1.scatter(data=adata_files['adata'+i].var, x='means',y=\"dispersions\",\n",
    "                              c=c,s = s,cmap=cmap,linewidth=linewidth,rasterized=True)\n",
    "        ax1.set_xlabel('mean expression') \n",
    "        ax1.set_ylabel('dispersion') \n",
    "        ax1.legend(scatter.legend_elements()[0][0:2],['other genes','highly variable'],\n",
    "                   loc=\"upper right\",)\n",
    "        scatter = ax2.scatter(data=adata_files['adata'+i].var, x='means',y=\"dispersions_norm\",\n",
    "                              c=c,s = s,cmap=cmap,linewidth=linewidth,rasterized=True)\n",
    "        ax2.legend(scatter.legend_elements()[0][0:2],['other genes','highly variable'],\n",
    "                   loc=\"upper right\",)\n",
    "        ax2.set_xlabel('mean expression') \n",
    "        ax2.set_ylabel('dispersions_norm')\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##PCA variance plot\n",
    "        sc.pl.pca_variance_ratio(adata_files['adata'+i], log=True, show=False)\n",
    "        pdf.savefig()\n",
    "        plt.close()\n",
    "        \n",
    "        ##umaps\n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files['adata'+i],\n",
    "                        color=['doublet_score','leiden','total_counts','pct_counts_mt','pct_counts_rb','pct_counts_hb','phase',]+genes_to_plot,\n",
    "                        show = False,ncols = 2)\n",
    "        for p in ax:\n",
    "            p.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##Most highly expressed genes\n",
    "        plt.rcParams['figure.figsize'] = [8,4]\n",
    "        adata_files['adata'+i].X = adata_files['adata'+i].layers[\"original_counts\"].copy() #set raw counts to X because the layer argument is broken\n",
    "        ax = sc.pl.highest_expr_genes(adata_files['adata'+i], \n",
    "                                      gene_symbols = 'gene_name',\n",
    "                                      n_top=20,log=True,show=False)\n",
    "        ax.set_rasterized(True)\n",
    "        pdf.savefig()\n",
    "        plt.close()\n",
    "        ##return normalised counts to X\n",
    "        adata_files['adata'+i].X = adata_files['adata'+i].layers[\"normalized_counts\"].copy()\n",
    "\n",
    "        ##doublet scatterplots\n",
    "        ###Scatterplots\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap=scatter_cmap\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files['adata'+i].obs, x=x1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax2.scatter(data=adata_files['adata'+i].obs, x=x2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax3.scatter(data=adata_files['adata'+i].obs, x=y1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax3.set_xlabel(y1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax4.scatter(data=adata_files['adata'+i].obs, x=y2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax4.set_xlabel(y2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        fig.colorbar(mpl.cm.ScalarMappable(norm=mpl.colors.Normalize(0, max(adata_files['adata'+i].obs[c])), cmap=cmap),\n",
    "                     ax=ax1, orientation='vertical', label=c)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "        \n",
    "        #boxplot of doublet scores\n",
    "        ax = sns.boxplot(data=adata_files['adata'+i].obs, x=\"leiden\", y=\"doublet_score\",\n",
    "            palette = adata_files['adata'+i].uns['leiden_colors'], \n",
    "                 fliersize=0)\n",
    "        cnum = 0\n",
    "        for line in ax.lines:\n",
    "            if line.get_linestyle() == 'None':\n",
    "                sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                              size = 1.5, color = adata_files['adata'+i].uns['leiden_colors'][cnum], rasterized = True)\n",
    "                cnum = cnum+1\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "    ##saving h5ad\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Saving data to h5ad file')\n",
    "    !mkdir -p {'h5ad_files/'+species+'/'+genome+'/'}\n",
    "    adata_files['adata'+i].write('h5ad_files/'+species+'/'+genome+'/'+i+'_before_filtering.h5ad')\n",
    "    \n",
    "\n",
    "    ##remove used variables from memory\n",
    "    del adata_cellr\n",
    "    del adata_veloc\n",
    "    del adata_merge\n",
    "    gc.collect()\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Loaded files for ',i)\n",
    "print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Finished Reading file(s)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d634e27d-d778-48ee-b8ab-7772f7679d5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "##can I filter these objects with the barcodes from mm10?\n",
    "adata_filt = sc.read('h5ad_files/mouse/mm10/adata_concat.h5ad')\n",
    "for i in sample_names:\n",
    "    adata_files['adata'+i] = adata_files['adata'+i][adata_filt[adata_filt.obs.index.isin(adata_files['adata'+i].obs.index)].obs.index]\n",
    "    adata_files['adata'+i].write('h5ad_files/'+species+'/'+genome+'/'+i+'_filtered_by_mm10.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "431075ab-24be-48a7-b42f-0e40697cfcbd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy_2025",
   "language": "python",
   "name": "scanpy_2025"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
