{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "16057746-ec86-443d-ab8b-cbde474a7d99",
   "metadata": {},
   "source": [
    "# Notes on QC  \n",
    "Initially I filtered cells based on QC variables.  \n",
    "Then I reassessed the filtered cells and ran scrublet for doublet simulation.\n",
    "\n",
    "I will filter both with cutoffs and clusters. Sometimes it's difficult to design cutoffs that capture all the bad cells.  \n",
    "Clustering will often group them together well. If when assessing a sample I find a cluster of what I percieve as entirely bad cells, I will remove the cluster.  \n",
    "## SE8_9\n",
    "Looking at SE8_9 I chose the following filtering parameters:  \n",
    "'min_counts':8000, 'max_counts':100000, 'min_genes':2500, 'max_genes':15000,  'min_mito':0,  'max_mito':5,   'min_rb':7.5,  'max_rb':25  \n",
    "\n",
    "In our scatter and boxplots we can see 3 clusters that fall outside the filtering parameters.  \n",
    "Cluster 0, 8 and 12  \n",
    "\n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE8_9_scatter_qc.png)  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE8_9_boxplot.png)  \n",
    "\n",
    "I want to check what they are before removing them.  \n",
    "\n",
    "Cluster 0 and 8 -> poor quality cells  \n",
    "Cluster 12 -> Just seems to have some cells with high mito, I will filter the data by mito but not outright remove this cluster  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE8_9_umap.png)  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE8_9_QC_umap.png)  \n",
    "\n",
    "\n",
    "I will mark the bad clusters as failed.:  \n",
    "['0','8']  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "Here cluster 25 stands out as being composed of likely doublets. In addition we can see small clumps of doublets around.  \n",
    "I filtered all cells by a doublet score of 0.15 and removed cluster 25.  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE8_9_pass_umap.png)   \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE8_9_pass_doublet2_umap.png)   \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE8_9_pass_doublet_boxplot.png)   \n",
    "\n",
    "## SE10\n",
    "Looking at SE10 I chose the following filtering parameters:  \n",
    "'min_counts':7000, 'max_counts':100000, 'min_genes':2500, 'max_genes':15000,  'min_mito':0,  'max_mito':7.5,   'min_rb':4.5,  'max_rb':25  \n",
    "\n",
    "In our scatter and boxplots we can see  clusters that fall outside the filtering parameters.  \n",
    "Cluster 0  \n",
    "\n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE10_scatter_qc.png)  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE10_boxplot.png)  \n",
    "\n",
    "I want to check what they are before removing them.  \n",
    "\n",
    "Cluster 0 -> poor quality cells  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE10_umap.png)  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE10_QC_umap.png)  \n",
    "\n",
    "\n",
    "I will mark the bad clusters as failed.:  \n",
    "['0',]  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "Here cluster 6 stands out as being composed of likely doublets. In addition we can see small clumps of doublets around.  \n",
    "I filtered all cells by a doublet score of 0.15 and removed cluster 6.  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE10_pass_umap.png)   \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE10_pass_doublet2_umap.png)   \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE10_pass_doublet_boxplot.png)  \n",
    "\n",
    "\n",
    "## SE11\n",
    "Looking at SE11 I chose the following filtering parameters:  \n",
    "'min_counts':3500, 'max_counts':100000, 'min_genes':1500, 'max_genes':15000,  'min_mito':0,  'max_mito':9,   'min_rb':4,  'max_rb':25  \n",
    "\n",
    "In our scatter and boxplots we can see  clusters that fall outside the filtering parameters.  \n",
    "Cluster 0 is trash cells  \n",
    "Cluster 17, 18, 19 and 20 partially fall outside these parameters. I will look at them more closely  \n",
    "\n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE11_scatter_qc.png)  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE11_boxplot.png)  \n",
    "\n",
    "I want to check what they are before removing them.  \n",
    "\n",
    "Cluster 0 -> poor quality cells  \n",
    "Clusters 17, 18, 19 and 20 all have unique marker genes. They seem to just have a portion of cells that are of low quality.  \n",
    "I will keep the clusters but remove the cells outside of the threshold values  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE11_umap.png)  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE11_QC_umap.png)  \n",
    "\n",
    "\n",
    "I will mark the bad clusters as failed.:  \n",
    "['0',]  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "Here cluster 2 stands out as being composed of likely doublets. In addition we can see small clumps of doublets around. The doublet score for cluster 2 is not incredibly high, but looking at its DEGs it doesn't really look like anything. It's strict, but I will remove these cells.\n",
    "I filtered all cells by a doublet score of 0.12 and removed cluster 2.  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE11_pass_umap.png)   \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE11_pass_doublet2_umap.png)   \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE11_pass_doublet_boxplot.png)  \n",
    "\n",
    "## SE12\n",
    "Looking at SE12 I chose the following filtering parameters:  \n",
    "'min_counts':3500, 'max_counts':100000, 'min_genes':1500, 'max_genes':15000,  'min_mito':0,  'max_mito':10,   'min_rb':4,  'max_rb':25  \n",
    "\n",
    "In our scatter and boxplots we can see  clusters that fall outside the filtering parameters.    \n",
    "Cluster 1 is trash cells  \n",
    "Cluster 6 and 8 partially fall outside these parameters. I will look at them more closely  \n",
    "\n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE12_scatter_qc.png)  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE12_boxplot.png)  \n",
    "\n",
    "I want to check what they are before removing them.  \n",
    "\n",
    "Cluster 1 -> poor quality cells    \n",
    "Clusters 6 and 8 are a little odd. They seem to be brain related, but it's difficult to find extremely unique markers. I wonder if they are somehow stressed cells.  \n",
    "For now I will keep these clusters but remove the cells outside of the threshold values. Perhaps they will still be removed later.  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE12_umap.png)  \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE12_QC_umap.png)  \n",
    "\n",
    "\n",
    "I will mark the bad clusters as failed.:  \n",
    "['1',]  \n",
    "\n",
    "Keeping the cells that passed the QC, I ran scrublet for doublet simulation.  \n",
    "Here cluster 24 and 43 at highe resolution stands out as being composed of likely doublets. In addition we can see small clumps of doublets around. \n",
    "I filtered all cells by a doublet score of 0.13 and removed cluster 24 and 43. \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE12_pass_umap.png)   \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE12_pass_doublet2_umap.png)   \n",
    "![](../markdown_images/catshark/sScyCan1.1/markdown_plots/SE12_pass_doublet_boxplot.png)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3217d80e-06a5-47dd-a923-5e72c95bd076",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.getLogger('matplotlib.font_manager').setLevel(logging.ERROR)\n",
    "import scanpy as sc\n",
    "import scvelo as scv\n",
    "import scvi\n",
    "import seaborn as sns\n",
    "import plotly.express as px\n",
    "import numpy as np\n",
    "from dash import Dash, dcc, html, Input, Output\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import warnings\n",
    "import logging\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "import gc\n",
    "os.environ['R_HOME'] = sys.exec_prefix+\"/lib/R/\"\n",
    "\n",
    "# Plotting\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "from matplotlib.colors import LinearSegmentedColormap, ListedColormap\n",
    "from matplotlib.lines import Line2D \n",
    "\n",
    "import logging\n",
    "logger = logging.getLogger()\n",
    "logger.setLevel(logging.CRITICAL)\n",
    "\n",
    "from copy import copy\n",
    "reds = copy(mpl.cm.Reds)\n",
    "reds.set_under(\"lightgray\")\n",
    "\n",
    "project_directory = '/Cranio_Lab/Louk_Seton/4_species_project'\n",
    "os.chdir(os.path.expanduser(\"~\")+project_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5d2271a-0e86-4ac8-9e38-2ecd59e4da15",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gtfparse breaks attribute expansion, need to do it myself\n",
    "def attribute_dict(attributes):\n",
    "    tmp_dict = {}\n",
    "    try: \n",
    "        for i in range(len(list(filter(None,list(attributes))))):\n",
    "            tmp_dict[list(attributes)[i].lstrip(' ').split(\" \", 1)[0]] = list(attributes)[i].lstrip(' ').split(\" \", 1)[1].strip('\\'')\n",
    "    except: \n",
    "        print('no attributes found')        \n",
    "    return tmp_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "460021f8-62e2-4b1f-aea3-33ce993d443f",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "from gtfparse import parse_gtf_pandas\n",
    "df = parse_gtf_pandas('required_files/shark_genes.gtf')\n",
    "#first create a dict by applying the function to the column. \n",
    "#then expand dict to columns\n",
    "df['attribute_dict'] = df['attribute_split'].apply(attribute_dict)\n",
    "df = pd.concat([df, df['attribute_dict'].apply(pd.Series)], axis=1)\n",
    "df = df.iloc[:,[0,1,2,6,11,12,14,19,20]] #keep only relevant columns\n",
    "#subset cds, these have the product attribute\n",
    "df_cds = df.loc[df['feature'] == 'CDS']\n",
    "df_cds = df_cds.drop_duplicates(subset = \"gene_id\")\n",
    "df_cds = df_cds.set_index('gene_id')\n",
    "#subset genes, because not all genes have a cds \n",
    "df_genes = df.loc[df['feature'] == 'gene']\n",
    "df_genes = df_genes.set_index('gene_id')\n",
    "#join the two, so that genes now have a product attribute\n",
    "df_products = df_genes.iloc[:,[0,1,2,3,4,5]].join(df_cds.iloc[:,[6,7]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef50b89e-7bfa-48db-9396-8eb483fa4598",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "##TODO\n",
    "\n",
    "######shark calcification######\n",
    "start_time=time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\")\n",
    "print('start time:',start_time)\n",
    "\n",
    "sample_names = ['SE8_9','SE10','SE11','SE12',] #specify the sample names\n",
    "species = 'catshark' #specify the species\n",
    "genome = 'sScyCan1.1' #specify the genome\n",
    "output_prefix = 'cellranger_related/cellranger_output/sc_rnaseq/' #specify the location of the cellranger output\n",
    "\n",
    "#shark stuff\n",
    "mt_genes = ['ND1','ND2','ND3','ND4','ND4L','ND5','ND6','COX1','COX2','COX3','ATP6','ATP8','CYTB']\n",
    "\n",
    "\n",
    "##plotting vars\n",
    "output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "genes_to_plot = ['sox10',]\n",
    "dot_size = .6\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "##################################################################################################################\n",
    "cellranger_locs = []\n",
    "velocyto_locs = []\n",
    "\n",
    "for sample in sample_names: #simple loop to create a list with the location of the required files\n",
    "    cellranger_locs = cellranger_locs + [output_prefix+species+'/'+genome+'/'+sample+'/outs/filtered_feature_bc_matrix/']\n",
    "    velocyto_locs = velocyto_locs + [output_prefix+species+'/'+genome+'/'+sample+'/velocyto/'+sample+'.loom']\n",
    "\n",
    "warnings.filterwarnings('ignore')\n",
    "logging.getLogger().setLevel(logging.CRITICAL)\n",
    "\n",
    "adata_files={}\n",
    "print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Reading file(s)')\n",
    "for i,j,k in zip(sample_names, cellranger_locs, velocyto_locs):\n",
    "    ##Read and merge 10x and velocyto output, preserving gene names and ensemblid\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Sample ',i)\n",
    "    adata_cellr = sc.read_10x_mtx(j)\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Loaded 10x mtx')\n",
    "    adata_cellr.obs.index = adata_cellr.obs.index.str[:-2]\n",
    "   \n",
    "    adata_veloc = sc.read(k)\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Loaded velocyto loom')\n",
    "    adata_veloc.obs.index = adata_veloc.obs.index.str.split(':').str[1]\n",
    "    adata_veloc.obs.index = adata_veloc.obs.index.str[:-1]\n",
    "    adata_veloc.var.index = adata_veloc.var['Accession']\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Merging data')\n",
    "    adata_merge = scv.utils.merge(adata_cellr, adata_veloc)\n",
    "    adata_files[i]=adata_merge\n",
    "    adata_files[i].var = adata_files[i].var.join(df_products.iloc[:,[1,4,6,7]])\n",
    "    \n",
    "    adata_files[i].obs['sample']=i\n",
    "    adata_files[i].obs['ref_genome']=genome\n",
    "    adata_files[i].obs['barcode']=adata_files[i].obs.index\n",
    "    adata_files[i].obs.index=adata_files[i].obs['barcode']+'.'+adata_files[i].obs['sample']\n",
    "\n",
    "    adata_files[i].var.index = list(adata_files[i].var['gene_ids'])#set gene_ids as index\n",
    "\n",
    "    ##Calculate QC values\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Calculating QC values for cells')\n",
    "    adata_files[i].var['mt'] = adata_files[i].var['gene_ids'].isin(mt_genes)  # annotate the group of mitochondrial genes as 'mt'\n",
    "    adata_files[i].var['rb'] = adata_files[i].var['gene_ids'].str.startswith(('rps','rpl'))  # annotate the group of ribosomal proteins as 'rb'\n",
    "\n",
    "    adata_files[i].var['hb'] = adata_files[i].var['product'].str.startswith(('hemoglobin subunit'),na=False) #check this\n",
    "    adata_files[i].var_names_make_unique()\n",
    "    adata_files[i].obs_names_make_unique()\n",
    "    sc.pp.calculate_qc_metrics(adata_files[i], qc_vars=[\"mt\", \"rb\", \"hb\"], inplace=True, log1p=True)\n",
    "    #clip to meaningful numbers, better for comparisons\n",
    "    adata_files[i].obs['pct_counts_mt'] = adata_files[i].obs['pct_counts_mt'].clip(upper = 20)\n",
    "   # adata_files[i].obs['pct_counts_hb'] = adata_files[i].obs['pct_counts_hb'].clip(upper = 20)\n",
    "    #adata_files[i].obs['pct_counts_rb'] = adata_files[i].obs['pct_counts_rb'].clip(upper = 40)\n",
    "    adata_files[i].obs['total_counts'] = adata_files[i].obs['total_counts'].clip(upper = 50000)\n",
    "\n",
    "    ##calculate doublet score\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Running scrublet for doublet detection')\n",
    "    sc.pp.scrublet(adata_files[i])\n",
    "\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Initial preprocess and clustering')\n",
    "    ##standard normalization\n",
    "    adata_files[i].layers[\"original_counts\"] = adata_files[i].X.copy() #store original counts\n",
    "    sc.pp.normalize_total(adata_files[i]) # Normalizing to median total counts\n",
    "    sc.pp.log1p(adata_files[i]) # Logarithmize the data\n",
    "    adata_files[i].layers[\"normalized_counts\"] = adata_files[i].X.copy()\n",
    "\n",
    "    ##highly variable genes\n",
    "    sc.pp.highly_variable_genes(adata_files[i], n_top_genes=2000,)\n",
    "\n",
    "    ##dimensionality reduction and clustering\n",
    "    sc.tl.pca(adata_files[i])\n",
    "    sc.pp.neighbors(adata_files[i])\n",
    "    sc.tl.umap(adata_files[i])\n",
    "    sc.tl.leiden(adata_files[i],)\n",
    "\n",
    "    ##cell cycle scoring\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Prepare cell cycle scoring')\n",
    "    cell_cycle_genes = [x.strip() for x in open('required_files/shark_cell_cycle.txt')]\n",
    "    s_genes = cell_cycle_genes[:43]\n",
    "    s_genes = [x for x in s_genes if x in list(adata_files[i].var['gene_ids'])]\n",
    "    g2m_genes = cell_cycle_genes[43:]\n",
    "    g2m_genes = [x for x in g2m_genes if x in list(adata_files[i].var['gene_ids'])]\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Cell cycle scoring')\n",
    "    sc.tl.score_genes_cell_cycle(adata_files[i], s_genes=s_genes, g2m_genes=g2m_genes)\n",
    "\n",
    "    \n",
    "    #plot pdfs of qc\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Plotting QC and saving to PDF')\n",
    "    with PdfPages(output_dir+i+'_QC.pdf') as pdf:\n",
    "        mpl.rcParams['axes.grid'] = False\n",
    "        ax = sc.pl.violin(adata_files[i],[\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\",\"pct_counts_hb\", \"pct_counts_rb\",'doublet_score'],\n",
    "                          stripplot=False, multi_panel=True,show=False,\n",
    "                          height=4, aspect=.8,col_wrap = 3)\n",
    "        plt.suptitle(i, x = 0.1)\n",
    "        pdf.savefig()  # saves the current figure into a pdf page\n",
    "        plt.close() # close the plot and page\n",
    "\n",
    "        ###Scatterplots\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap=scatter_cmap\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files[i].obs, x=x1,y=y1,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax1.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax1.set_ylabel(y1)\n",
    "        ax2.scatter(data=adata_files[i].obs, x=x2,y=y1,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax2.set_xlim(adata_files[i].obs[x2].quantile(0) ,adata_files[i].obs[x2].quantile(0.95))\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax2.set_ylabel(y1)\n",
    "        ax3.scatter(data=adata_files[i].obs, x=x1,y=y2,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax3.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "        ax3.set_xlabel(x1) \n",
    "        ax3.set_ylabel(y2)\n",
    "        ax4.scatter(data=adata_files[i].obs, x=x2,y=y2,\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax4.set_xlim(adata_files[i].obs[x2].quantile(0) ,adata_files[i].obs[x2].quantile(0.95))\n",
    "        ax4.set_xlabel(x2) \n",
    "        ax4.set_ylabel(y2)\n",
    "\n",
    "\n",
    "        fig.colorbar(mpl.cm.ScalarMappable(norm=mpl.colors.Normalize(0, max(adata_files[i].obs[c])), cmap=cmap),\n",
    "                     ax=ax1, orientation='vertical', label=c)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##nice jointplot shows good cells\n",
    "        mpl.rcParams['axes.grid'] = False\n",
    "        sns_plot = sns.jointplot(data=adata_files[i].obs,\n",
    "                                 x=\"log1p_total_counts\",\n",
    "                                 y=\"log1p_n_genes_by_counts\",\n",
    "                                 kind=\"hex\",\n",
    "                                 marginal_kws=dict(bins=100,),\n",
    "                                 joint_kws = dict(gridsize=50,),\n",
    "                                )\n",
    "        sns_plot.fig.suptitle(i,x=0.05)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##custom hvg plot\n",
    "        fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize = (8,4))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap = ListedColormap(['grey','black'])\n",
    "        c = 'highly_variable'\n",
    "        scatter = ax1.scatter(data=adata_files[i].var, x='means',y=\"dispersions\",\n",
    "                              c=c,s = s,cmap=cmap,linewidth=linewidth,rasterized=True)\n",
    "        ax1.set_xlabel('mean expression') \n",
    "        ax1.set_ylabel('dispersion') \n",
    "        ax1.legend(scatter.legend_elements()[0][0:2],['other genes','highly variable'],\n",
    "                   loc=\"upper right\",)\n",
    "        scatter = ax2.scatter(data=adata_files[i].var, x='means',y=\"dispersions_norm\",\n",
    "                              c=c,s = s,cmap=cmap,linewidth=linewidth,rasterized=True)\n",
    "        ax2.legend(scatter.legend_elements()[0][0:2],['other genes','highly variable'],\n",
    "                   loc=\"upper right\",)\n",
    "        ax2.set_xlabel('mean expression') \n",
    "        ax2.set_ylabel('dispersions_norm')\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##PCA variance plot\n",
    "        sc.pl.pca_variance_ratio(adata_files[i], log=True, show=False)\n",
    "        pdf.savefig()\n",
    "        plt.close()\n",
    "        \n",
    "        ##umaps\n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files[i],\n",
    "                        color=['doublet_score','leiden','total_counts','pct_counts_mt','pct_counts_rb','pct_counts_hb','phase',]+genes_to_plot,\n",
    "                        show = False,ncols = 2)\n",
    "        for p in ax:\n",
    "            p.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##Most highly expressed genes\n",
    "        plt.rcParams['figure.figsize'] = [8,4]\n",
    "        adata_files[i].X = adata_files[i].layers[\"original_counts\"].copy() #set raw counts to X because the layer argument is broken\n",
    "        ax = sc.pl.highest_expr_genes(adata_files[i], \n",
    "                                      gene_symbols = 'gene_ids',\n",
    "                                      n_top=20,log=True,show=False)\n",
    "        ax.set_rasterized(True)\n",
    "        pdf.savefig()\n",
    "        plt.close()\n",
    "        ##return normalised counts to X\n",
    "        adata_files[i].X = adata_files[i].layers[\"normalized_counts\"].copy()\n",
    "\n",
    "        ##doublet scatterplots\n",
    "        ###Scatterplots\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap=scatter_cmap\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files[i].obs, x=x1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax2.scatter(data=adata_files[i].obs, x=x2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax3.scatter(data=adata_files[i].obs, x=y1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax3.set_xlabel(y1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax4.scatter(data=adata_files[i].obs, x=y2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax4.set_xlabel(y2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        fig.colorbar(mpl.cm.ScalarMappable(norm=mpl.colors.Normalize(0, max(adata_files[i].obs[c])), cmap=cmap),\n",
    "                     ax=ax1, orientation='vertical', label=c)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "        \n",
    "        #boxplot of doublet scores\n",
    "        ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"doublet_score\",\n",
    "            palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0)\n",
    "        cnum = 0\n",
    "        for line in ax.lines:\n",
    "            if line.get_linestyle() == 'None':\n",
    "                sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                              size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True)\n",
    "                cnum = cnum+1\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "    ##saving h5ad\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Saving data to h5ad file')\n",
    "    !mkdir -p {'h5ad_files/'+species+'/'+genome}\n",
    "    adata_files[i].write('h5ad_files/'+species+'/'+genome+'/'+i+'_before_filtering.h5ad')\n",
    "    \n",
    "\n",
    "    ##remove used variables from memory\n",
    "    del adata_cellr\n",
    "    del adata_veloc\n",
    "    del adata_merge\n",
    "    gc.collect()\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Loaded files for ',i)\n",
    "print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Finished Reading file(s)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "876d0400-6439-4af4-99b1-4856e9177bec",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "##Filter using QC vars and make plots\n",
    "filter_vars = {'SE8_9':{'min_counts':8000, 'max_counts':100000, 'min_genes':2500, 'max_genes':15000,  'min_mito':0,  'max_mito':5,   'min_rb':7.5,  'max_rb':25,\n",
    "                      'clusters':['0','8']},\n",
    "               'SE10':{'min_counts':7000, 'max_counts':100000, 'min_genes':2500, 'max_genes':15000,  'min_mito':0,  'max_mito':7.5,   'min_rb':4.5,  'max_rb':25,\n",
    "                      'clusters':['0',]}, \n",
    "               'SE11':{'min_counts':3500, 'max_counts':100000, 'min_genes':1500, 'max_genes':15000,  'min_mito':0,  'max_mito':9,   'min_rb':4,  'max_rb':25,\n",
    "                      'clusters':['0']},\n",
    "               'SE12':{'min_counts':7000, 'max_counts':100000, 'min_genes':2000, 'max_genes':15000,  'min_mito':0,  'max_mito':7.5,   'min_rb':5,  'max_rb':35,\n",
    "                      'clusters':['1']},}\n",
    "\n",
    "filter_vars = pd.DataFrame(filter_vars).T\n",
    "\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "dot_size = 1\n",
    "linewidth = 0\n",
    "scatter_color = 'QC'\n",
    "output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "for sample in list(filter_vars.index):\n",
    "    min_counts = filter_vars.loc[sample,'min_counts']\n",
    "    max_counts = filter_vars.loc[sample,'max_counts']\n",
    "    min_genes = filter_vars.loc[sample,'min_genes']\n",
    "    max_genes = filter_vars.loc[sample,'max_genes']\n",
    "    min_mito = filter_vars.loc[sample,'min_mito']\n",
    "    max_mito = filter_vars.loc[sample,'max_mito']\n",
    "    min_rb = filter_vars.loc[sample,'min_rb']\n",
    "    max_rb = filter_vars.loc[sample,'max_rb']\n",
    "\n",
    "    conditions = [\n",
    "    (adata_files[sample].obs['n_genes_by_counts'] < min_genes),\n",
    "    (adata_files[sample].obs['n_genes_by_counts'] > max_genes),\n",
    "    (adata_files[sample].obs['total_counts'] < min_counts),\n",
    "    (adata_files[sample].obs['total_counts'] > max_counts),\n",
    "    (adata_files[sample].obs['pct_counts_mt'] > max_mito),\n",
    "    (adata_files[sample].obs['pct_counts_mt'] < min_mito),\n",
    "    (adata_files[sample].obs['pct_counts_rb'] < min_rb),\n",
    "    (adata_files[sample].obs['pct_counts_rb'] > max_rb),\n",
    "    (adata_files[sample].obs['pct_counts_mt'] <= max_mito) & (adata_files[sample].obs['pct_counts_mt'] >= min_mito) & (adata_files[sample].obs['n_genes_by_counts'] >= min_genes) & (adata_files[sample].obs['n_genes_by_counts'] <= max_genes) & \n",
    "    (adata_files[sample].obs['total_counts'] >= min_counts) & (adata_files[sample].obs['total_counts'] <= max_counts) & (adata_files[sample].obs['pct_counts_rb'] >= min_rb) & (adata_files[sample].obs['pct_counts_rb'] <= max_rb)\n",
    "    ]\n",
    "    \n",
    "    values = ['Low_nFeature', 'High_nFeature', 'Low_Counts','High_Counts', 'High_MT','Low_MT', 'Low_RB' ,'High_RB' ,'Pass']\n",
    "    adata_files[sample].obs['QC'] = np.select(conditions, values)\n",
    "    adata_files[sample].obs.loc[adata_files[sample].obs['leiden'].isin(filter_vars.loc[sample,'clusters']), 'QC'] = \"Bad_Cluster\"\n",
    "    adata_files[sample].obs['QC'] = adata_files[sample].obs['QC'].astype('category')\n",
    "\n",
    "    colors = {'Low_nFeature':'#d9d9d9', \n",
    "                  'High_nFeature':'#8dd3c7', \n",
    "                  'Low_Counts':'#ffffb3',\n",
    "                  'High_Counts':'#bebada',\n",
    "                  'High_MT':'#fb8072',\n",
    "                  'Low_MT':'#80b1d3',\n",
    "                  'Low_RB':'#fdb462',\n",
    "                  'High_RB':'#fccde5',\n",
    "                  'Bad_Cluster':'#000000',\n",
    "                  'Pass':'#b3de69',}\n",
    "    with PdfPages(output_dir+sample+'_filter_params.pdf') as pdf:        \n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files[sample], color = ['QC'],palette = colors, show=False)\n",
    "        ax.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150,bbox_inches='tight')  # saves the current figure into a pdf page\n",
    "        plt.close() # close the plot and page\n",
    "        \n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files[sample], color = ['n_genes_by_counts','doublet_score','phase','pct_counts_hb'],ncols=2,show=False)\n",
    "        for p in ax:\n",
    "            p.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150)  # saves the current figure into a pdf page\n",
    "        plt.close() # close the plot and page\n",
    "\n",
    "        ###Scatterplots\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files[sample].obs, x=x1,y=y1,\n",
    "                    c = adata_files[sample].obs[c].map(colors),s = s,linewidth=linewidth,rasterized=True)\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax1.set_ylabel(y1)\n",
    "        ax2.scatter(data=adata_files[sample].obs, x=x2,y=y1,\n",
    "                    c = adata_files[sample].obs[c].map(colors),s = s,linewidth=linewidth,rasterized=True)\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax2.set_ylabel(y1)\n",
    "        ax3.scatter(data=adata_files[sample].obs, x=x1,y=y2,\n",
    "                    c = adata_files[sample].obs[c].map(colors),s = s,linewidth=linewidth,rasterized=True)\n",
    "        ax3.set_xlabel(x1) \n",
    "        ax3.set_ylabel(y2)\n",
    "        scatter = ax4.scatter(data=adata_files[sample].obs, x=x2,y=y2,\n",
    "                    c = adata_files[sample].obs[c].map(colors),s = s,linewidth=linewidth,rasterized=True)\n",
    "        ax4.set_xlabel(x2) \n",
    "        ax4.set_ylabel(y2)\n",
    "        handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=8) for k, v in colors.items()]\n",
    "        ax4.legend(title='color', handles=handles, bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bafaa74-9fd0-4445-beb8-0cf5177a5bc5",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "##doublet stuff and some more plots\n",
    "genes_to_plot = ['sox10','shha','paplna','LOC119973325','thbs4a','mpz','matn1','col9a2','epyc','postnb',\n",
    "                 'hydin','sox2','myrf','nrxn1a','myo3b','th','vegfc','scube3',]\n",
    "dot_size = 1\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "\n",
    "for i in sample_names:\n",
    "    adata_files[i+'_pass'] = adata_files[i][adata_files[i].obs['QC']=='Pass'].copy() #filter cells\n",
    "    i = i+'_pass'\n",
    "    adata_files[i].X = adata_files[i].layers[\"original_counts\"].copy() #restore original counts\n",
    "    sc.pp.scrublet(adata_files[i])\n",
    "\n",
    "    print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Initial preprocess and clustering')\n",
    "    ##standard normalization\n",
    "    sc.pp.normalize_total(adata_files[i]) # Normalizing to median total counts\n",
    "    sc.pp.log1p(adata_files[i]) # Logarithmize the data\n",
    "    adata_files[i].layers[\"normalized_counts\"] = adata_files[i].X.copy()\n",
    "\n",
    "    ##highly variable genes\n",
    "    sc.pp.highly_variable_genes(adata_files[i], n_top_genes=2000,)\n",
    "\n",
    "    ##dimensionality reduction and clustering\n",
    "    sc.tl.pca(adata_files[i])\n",
    "    sc.pp.neighbors(adata_files[i])\n",
    "    sc.tl.umap(adata_files[i])\n",
    "    sc.tl.leiden(adata_files[i],resolution = 1, key_added = 'leiden_filt')\n",
    "    sc.tl.leiden(adata_files[i],resolution = 2, key_added = 'leiden_filt_highres')\n",
    "\n",
    "    ##plotting new stuff\n",
    "    output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "    !mkdir -p {output_dir}\n",
    "\n",
    "    with PdfPages(output_dir+i+'_doublet_check.pdf') as pdf:\n",
    "        ##umaps\n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        ax = sc.pl.umap(adata_files[i],\n",
    "                        color=['doublet_score','leiden','total_counts','leiden_filt','pct_counts_mt','leiden_filt_highres','pct_counts_rb','pct_counts_hb','phase',]+genes_to_plot,\n",
    "                        show = False,ncols = 2)\n",
    "        for p in ax:\n",
    "            p.set_rasterized(True)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "        \n",
    "        #boxplot of doublet scores\n",
    "        plt.rcParams['figure.figsize'] = [8,4]\n",
    "        ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"doublet_score\",\n",
    "            palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "            palette = adata_files[i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "        cnum = 0\n",
    "        for line in ax.lines:\n",
    "            if line.get_linestyle() == 'None':\n",
    "                sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                              size = 1.5, color = adata_files[i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "                cnum = cnum+1\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt_highres\", y=\"doublet_score\",\n",
    "            palette = adata_files[i].uns['leiden_filt_highres_colors'], \n",
    "                 fliersize=0)\n",
    "        cnum = 0\n",
    "        for line in ax.lines:\n",
    "            if line.get_linestyle() == 'None':\n",
    "                sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                              size = 1.5, color = adata_files[i].uns['leiden_filt_highres_colors'][cnum], rasterized = True)\n",
    "                cnum = cnum+1\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()\n",
    "\n",
    "        ##doublet scatterplots\n",
    "        ###Scatterplots\n",
    "        plt.rcParams['figure.figsize'] = [5,4]\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2,layout='constrained',figsize=(10,8))\n",
    "        s=dot_size\n",
    "        linewidth=linewidth\n",
    "        cmap=scatter_cmap\n",
    "        c = scatter_color\n",
    "        ax1.scatter(data=adata_files[i].obs, x=x1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax1.set_xlabel(x1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax2.scatter(data=adata_files[i].obs, x=x2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax2.set_xlabel(x2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax3.scatter(data=adata_files[i].obs, x=y1,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax3.set_xlabel(y1) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        ax4.scatter(data=adata_files[i].obs, x=y2,y='doublet_score',\n",
    "                    c = c,s = s,linewidth=linewidth, cmap = cmap,rasterized=True)\n",
    "        ax4.set_xlabel(y2) \n",
    "        ax4.set_ylabel('doublet_score')\n",
    "        fig.colorbar(mpl.cm.ScalarMappable(norm=mpl.colors.Normalize(0, max(adata_files[i].obs[c])), cmap=cmap),\n",
    "                     ax=ax1, orientation='vertical', label=c)\n",
    "        pdf.savefig(dpi=150)\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d715172-318b-4f5d-a977-7e3e5712e871",
   "metadata": {},
   "outputs": [],
   "source": [
    "##removing doublets and generating data for analysis\n",
    "doublet_vars = {'SE8_9_pass':{'doublet_cutoff':.15, 'doublet_clusters':['25'],'doublet_clusters_high':None},\n",
    "                'SE10_pass':{'doublet_cutoff':.15, 'doublet_clusters':['6'],'doublet_clusters_high':None},\n",
    "                'SE11_pass':{'doublet_cutoff':.12, 'doublet_clusters':['2'],'doublet_clusters_high':None},\n",
    "                'SE12_pass':{'doublet_cutoff':.13, 'doublet_clusters':None,'doublet_clusters_high':['24','43']},\n",
    "               }\n",
    "\n",
    "doublet_vars = pd.DataFrame(doublet_vars).T\n",
    "\n",
    "genes_to_plot = ['shha','mpz','col9a2','postnb','sox2','hydin','nrxn1a','myo3b']\n",
    "\n",
    "for sample in list(doublet_vars.index):\n",
    "    if 'pass' in sample:\n",
    "        doublet_cutoff = doublet_vars.loc[sample,'doublet_cutoff']\n",
    "        doublet_clusters = doublet_vars.loc[sample,'doublet_clusters']\n",
    "        doublet_clusters_high = doublet_vars.loc[sample,'doublet_clusters_high']\n",
    "        adata_files[sample].obs['doublet_cutoff'] = 'pass'\n",
    "        adata_files[sample].obs.loc[adata_files[sample].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "        if isinstance(doublet_clusters, list):\n",
    "            adata_files[sample].obs.loc[adata_files[sample].obs['leiden_filt'].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "        if isinstance(doublet_clusters_high, list):\n",
    "            adata_files[sample].obs.loc[adata_files[sample].obs['leiden_filt_highres'].isin(doublet_clusters_high), 'doublet_cutoff'] = 'doublet'\n",
    "        adata_files[sample].obs['doublet_cutoff'] = adata_files[sample].obs['doublet_cutoff'].astype('category')\n",
    "\n",
    "        sn = sample+'_doublet'\n",
    "        adata_files[sn] = adata_files[sample][adata_files[sample].obs['doublet_cutoff'] == 'pass'].copy()\n",
    "        adata_files[sn].X = adata_files[sn].layers[\"original_counts\"].copy() #restore original counts\n",
    "        print(time.strftime(\"%Y_%m_%d-%I_%M_%S_%p\"),'Initial preprocess and clustering')\n",
    "        ##standard normalization\n",
    "        sc.pp.normalize_total(adata_files[sn]) # Normalizing to median total counts\n",
    "        sc.pp.log1p(adata_files[sn]) # Logarithmize the data\n",
    "        adata_files[sn].layers[\"normalized_counts\"] = adata_files[sn].X.copy()\n",
    "\n",
    "        ##highly variable genes\n",
    "        sc.pp.highly_variable_genes(adata_files[sn], n_top_genes=1000,)\n",
    "\n",
    "        ##dimensionality reduction and clustering\n",
    "        sc.tl.pca(adata_files[sn])\n",
    "        sc.pp.neighbors(adata_files[sn])\n",
    "        sc.tl.umap(adata_files[sn])\n",
    "        sc.tl.leiden(adata_files[sn],resolution = 1, key_added = 'leiden_post_QC')\n",
    "        sc.tl.leiden(adata_files[sn],resolution = 2, key_added = 'leiden_post_QC_highres')\n",
    "\n",
    "        ##plotting new stuff\n",
    "        output_dir = 'figures_publish/'+species+'/'+genome+'/QC_plots/'\n",
    "        !mkdir -p {output_dir}\n",
    "\n",
    "        i = sample\n",
    "        with PdfPages(output_dir+i+'_post_QC.pdf') as pdf:\n",
    "            ##umaps\n",
    "            plt.rcParams['figure.figsize'] = [5,4]\n",
    "            ax = sc.pl.umap(adata_files[sn],\n",
    "                            color=['leiden_post_QC','leiden_post_QC_highres','phase',]+genes_to_plot,\n",
    "                            show = False,ncols = 2)\n",
    "            for p in ax:\n",
    "                p.set_rasterized(True)\n",
    "            pdf.savefig(dpi=150)\n",
    "            plt.close()\n",
    "        \n",
    "        adata_files[sn].write('h5ad_files/'+species+'/'+genome+'/'+i+'_after_filtering.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb36cbdb-e931-4322-8bd4-3c040ddd6339",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "##markdown plots\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "## let's start with SE8_9\n",
    "i = 'SE8_9'\n",
    "filter_vars = {'SE8_9':{'min_counts':8000, 'max_counts':100000, 'min_genes':2500, 'max_genes':15000,  'min_mito':0,  'max_mito':5,   'min_rb':7.5,  'max_rb':25}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files[i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files[i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files[i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,10)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files[i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax2.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax2.set_ylim(0,20)\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplot of doublet scores\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden','total_counts','pct_counts_mt',],groups = ['0','8'], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "## SE10\n",
    "i = 'SE10'\n",
    "filter_vars = {'SE10':{'min_counts':7000, 'max_counts':100000, 'min_genes':2500, 'max_genes':15000,  'min_mito':0,  'max_mito':7.5,   'min_rb':4.5,  'max_rb':25}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files[i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files[i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files[i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,10)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files[i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax2.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax2.set_ylim(0,20)\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplot of doublet scores\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden','total_counts',],groups = ['0'], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "## SE11\n",
    "i = 'SE11'\n",
    "filter_vars = {'SE11':{'min_counts':3500, 'max_counts':100000, 'min_genes':1500, 'max_genes':15000,  'min_mito':0,  'max_mito':9,   'min_rb':4,  'max_rb':25}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files[i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files[i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files[i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,11)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files[i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax2.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax2.set_ylim(0,20)\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplot of doublet scores\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden','pct_counts_mt',],groups = ['0',], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "!mkdir -p {output_dir}\n",
    "\n",
    "## SE12\n",
    "i = 'SE12'\n",
    "filter_vars = {'SE12':{'min_counts':7000, 'max_counts':100000, 'min_genes':2000, 'max_genes':15000,  'min_mito':0,  'max_mito':7.5,   'min_rb':5,  'max_rb':35}, \n",
    "}   \n",
    "\n",
    "dot_size = 1.5\n",
    "linewidth = 0\n",
    "scatter_color = 'pct_counts_hb'\n",
    "scatter_cmap = 'coolwarm'\n",
    "y1 = 'pct_counts_mt' #these are for the QC scatter plots\n",
    "y2 = 'pct_counts_rb'\n",
    "x1 = 'n_genes_by_counts'\n",
    "x2 = 'total_counts'\n",
    "fig, ((ax1), (ax2)) = plt.subplots(1,2,layout='constrained',figsize=(8,4))\n",
    "s=dot_size\n",
    "linewidth=linewidth\n",
    "colors = dict(zip(list(adata_files[i].obs['leiden'].cat.categories), \n",
    "                  list(adata_files[i].uns['leiden_colors'])))\n",
    "\n",
    "ax1.scatter(data=adata_files[i].obs, x=x1,y=y1,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "ax1.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax1.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax1.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax1.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax1.set_xlabel(x1) \n",
    "ax1.set_ylabel(y1)\n",
    "ax1.set_ylim(0,11)\n",
    "handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=6) for k, v in colors.items()]\n",
    "ax1.legend(title='leiden', handles=handles, bbox_to_anchor=(.5, 1), loc='upper left',\n",
    "           ncol=3, fontsize = 9,labelspacing=0.2,columnspacing= .2,handletextpad=0)\n",
    "\n",
    "ax2.scatter(data=adata_files[i].obs, x=x1,y=y2,\n",
    "            s = s,linewidth=linewidth, \n",
    "            c = adata_files[i].obs['leiden'].map(colors),\n",
    "            rasterized=True)\n",
    "\n",
    "ax2.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "ax2.axvline(x=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axvline(x=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "ax2.set_xlim(adata_files[i].obs[x1].quantile(0) ,adata_files[i].obs[x1].quantile(0.95))\n",
    "ax2.set_ylim(0,20)\n",
    "ax2.set_xlabel(x1) \n",
    "ax2.set_ylabel(y2)\n",
    "plt.savefig(output_dir+i+'_scatter_qc.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "#boxplot of doublet scores\n",
    "fig, (ax1, ax2,ax3) = plt.subplots(3,1,layout='constrained',figsize=(8,8))\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_rb\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax1)\n",
    "ax1.axhline(y=filter_vars[i]['min_rb'],c = 'red', linewidth = .5)\n",
    "ax1.axhline(y=filter_vars[i]['max_rb'],c = 'red', linewidth = .5)\n",
    "\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax1)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"n_genes_by_counts\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax2)\n",
    "ax2.axhline(y=filter_vars[i]['min_genes'],c = 'red', linewidth = .5)\n",
    "ax2.axhline(y=filter_vars[i]['max_genes'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax2)\n",
    "        cnum = cnum+1\n",
    "\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden\", y=\"pct_counts_mt\",\n",
    "                 palette = adata_files[i].uns['leiden_colors'], \n",
    "                 fliersize=0, ax = ax3)\n",
    "ax3.axhline(y=filter_vars[i]['min_mito'],c = 'red', linewidth = .5)\n",
    "ax3.axhline(y=filter_vars[i]['max_mito'],c = 'red', linewidth = .5)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_colors'][cnum], rasterized = True, ax = ax3)\n",
    "        cnum = cnum+1\n",
    "\n",
    "plt.savefig(output_dir+i+'_boxplot.png', dpi = 80)\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden','pct_counts_mt','arhgef9a','syt1a','synpr'],groups = ['1','6','8'], ncols = 3,show = False)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "## post filter plots\n",
    "filter_vars = {'SE8_9':{'min_counts':8000, 'max_counts':100000, 'min_genes':2500, 'max_genes':15000,  'min_mito':0,  'max_mito':5,   'min_rb':7.5,  'max_rb':25,\n",
    "                      'clusters':['0','8']},\n",
    "               'SE10':{'min_counts':7000, 'max_counts':100000, 'min_genes':2500, 'max_genes':15000,  'min_mito':0,  'max_mito':7.5,   'min_rb':4.5,  'max_rb':25,\n",
    "                      'clusters':['0',]}, \n",
    "               'SE11':{'min_counts':3500, 'max_counts':100000, 'min_genes':1500, 'max_genes':15000,  'min_mito':0,  'max_mito':9,   'min_rb':4,  'max_rb':25,\n",
    "                      'clusters':['0']},\n",
    "               'SE12':{'min_counts':7000, 'max_counts':100000, 'min_genes':2000, 'max_genes':15000,  'min_mito':0,  'max_mito':7.5,   'min_rb':5,  'max_rb':35,\n",
    "                      'clusters':['1']},}\n",
    "\n",
    "filter_vars = pd.DataFrame(filter_vars).T\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "for sample in list(filter_vars.index):\n",
    "    min_counts = filter_vars.loc[sample,'min_counts']\n",
    "    max_counts = filter_vars.loc[sample,'max_counts']\n",
    "    min_genes = filter_vars.loc[sample,'min_genes']\n",
    "    max_genes = filter_vars.loc[sample,'max_genes']\n",
    "    min_mito = filter_vars.loc[sample,'min_mito']\n",
    "    max_mito = filter_vars.loc[sample,'max_mito']\n",
    "    min_rb = filter_vars.loc[sample,'min_rb']\n",
    "    max_rb = filter_vars.loc[sample,'max_rb']\n",
    "\n",
    "    conditions = [\n",
    "    (adata_files[sample].obs['n_genes_by_counts'] < min_genes),\n",
    "    (adata_files[sample].obs['n_genes_by_counts'] > max_genes),\n",
    "    (adata_files[sample].obs['total_counts'] < min_counts),\n",
    "    (adata_files[sample].obs['total_counts'] > max_counts),\n",
    "    (adata_files[sample].obs['pct_counts_mt'] > max_mito),\n",
    "    (adata_files[sample].obs['pct_counts_mt'] < min_mito),\n",
    "    (adata_files[sample].obs['pct_counts_rb'] < min_rb),\n",
    "    (adata_files[sample].obs['pct_counts_rb'] > max_rb),\n",
    "    (adata_files[sample].obs['pct_counts_mt'] <= max_mito) & (adata_files[sample].obs['pct_counts_mt'] >= min_mito) & (adata_files[sample].obs['n_genes_by_counts'] >= min_genes) & (adata_files[sample].obs['n_genes_by_counts'] <= max_genes) & \n",
    "    (adata_files[sample].obs['total_counts'] >= min_counts) & (adata_files[sample].obs['total_counts'] <= max_counts) & (adata_files[sample].obs['pct_counts_rb'] >= min_rb) & (adata_files[sample].obs['pct_counts_rb'] <= max_rb)\n",
    "    ]\n",
    "    \n",
    "    values = ['Fail', 'Fail', 'Fail','Fail', 'Fail','Fail', 'Fail' ,'Fail' ,'Pass']\n",
    "    adata_files[sample].obs['QC'] = np.select(conditions, values)\n",
    "    adata_files[sample].obs.loc[adata_files[sample].obs['leiden'].isin(filter_vars.loc[sample,'clusters']), 'QC'] = \"Fail\"\n",
    "    adata_files[sample].obs['QC'] = adata_files[sample].obs['QC'].astype('category')\n",
    "\n",
    "    plt.rcParams['figure.figsize'] = [4,3]\n",
    "    ax = sc.pl.umap(adata_files[sample], color = ['QC',], ncols = 3,show = False)\n",
    "    plt.savefig(output_dir+sample+'_QC_umap.png', dpi = 80,bbox_inches='tight')\n",
    "    plt.close()\n",
    "\n",
    "####Doublet stuff####\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "species = 'catshark' #specify the species\n",
    "genome = 'sScyCan1.1' #specify the genome\n",
    "#SE8_9\n",
    "i = 'SE8_9_pass'\n",
    "doublet_cutoff = .15\n",
    "doublet_clusters = ['25']\n",
    "doublet_filt_clust = 'leiden_filt'\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden_filt','leiden_filt_highres','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = doublet_cutoff,vmax = .4)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "##################################################################\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt_highres\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_highres_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_highres_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_highres_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files[i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files[i].obs.loc[adata_files[i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs.loc[adata_files[i].obs[doublet_filt_clust].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs['doublet_cutoff'] = adata_files[i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], size = 5,color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = 'doublet_cutoff', groups = ['doublet'],show = False)\n",
    "plt.savefig(output_dir+i+'_doublet2_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "####Doublet stuff####\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "species = 'catshark' #specify the species\n",
    "genome = 'sScyCan1.1' #specify the genome\n",
    "#SE10\n",
    "i = 'SE10_pass'\n",
    "doublet_cutoff = .15\n",
    "doublet_clusters = ['6']\n",
    "doublet_filt_clust = 'leiden_filt'\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden_filt','leiden_filt_highres','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = doublet_cutoff,vmax = .4)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "##################################################################\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt_highres\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_highres_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_highres_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_highres_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files[i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files[i].obs.loc[adata_files[i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs.loc[adata_files[i].obs[doublet_filt_clust].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs['doublet_cutoff'] = adata_files[i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], size = 5,color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = 'doublet_cutoff', groups = ['doublet'],show = False)\n",
    "plt.savefig(output_dir+i+'_doublet2_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "####Doublet stuff####\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "species = 'catshark' #specify the species\n",
    "genome = 'sScyCan1.1' #specify the genome\n",
    "#SE11\n",
    "i = 'SE11_pass'\n",
    "doublet_cutoff = .12\n",
    "doublet_clusters = ['2']\n",
    "doublet_filt_clust = 'leiden_filt'\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden_filt','leiden_filt_highres','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = doublet_cutoff,vmax = .4)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "##################################################################\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt_highres\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_highres_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_highres_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_highres_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files[i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files[i].obs.loc[adata_files[i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs.loc[adata_files[i].obs[doublet_filt_clust].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs['doublet_cutoff'] = adata_files[i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], size = 5,color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = 'doublet_cutoff', groups = ['doublet'],show = False)\n",
    "plt.savefig(output_dir+i+'_doublet2_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "####Doublet stuff####\n",
    "output_dir = 'markdown_images/'+species+'/'+genome+'/markdown_plots/'\n",
    "species = 'catshark' #specify the species\n",
    "genome = 'sScyCan1.1' #specify the genome\n",
    "#SE12\n",
    "i = 'SE12_pass'\n",
    "doublet_cutoff = .13\n",
    "doublet_clusters = ['24','43']\n",
    "doublet_filt_clust = 'leiden_filt_highres'\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = ['leiden_filt','leiden_filt_highres','doublet_score'],legend_loc = 'on data', ncols = 3,show = False, cmap = reds, vmin = doublet_cutoff,vmax = .4)\n",
    "plt.savefig(output_dir+i+'_umap.png', dpi = 80,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "##################################################################\n",
    "plt.rcParams['figure.figsize'] = [6,3]\n",
    "ax = sns.boxplot(data=adata_files[i].obs, x=\"leiden_filt_highres\", y=\"doublet_score\",\n",
    "                 palette = adata_files[i].uns['leiden_filt_highres_colors'], \n",
    "                 fliersize=0)\n",
    "cnum = 0\n",
    "for line in ax.lines:\n",
    "    if line.get_linestyle() == 'None':\n",
    "        sns.swarmplot(x=[int(b) for b in line.get_xdata()], y=line.get_ydata(),\n",
    "                      size = 1.5, color = adata_files[i].uns['leiden_filt_highres_colors'][cnum], rasterized = True)\n",
    "        cnum = cnum+1\n",
    "plt.savefig(output_dir+i+'_highres_doublet_boxplot.png',dpi=150, bbox_inches = 'tight')\n",
    "plt.close() \n",
    "\n",
    "adata_files[i].obs['doublet_cutoff'] = 'pass'\n",
    "adata_files[i].obs.loc[adata_files[i].obs['doublet_score'] > doublet_cutoff, 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs.loc[adata_files[i].obs[doublet_filt_clust].isin(doublet_clusters), 'doublet_cutoff'] = 'doublet'\n",
    "adata_files[i].obs['doublet_cutoff'] = adata_files[i].obs['doublet_cutoff'].astype('category')\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], size = 5,color = 'doublet_cutoff', show = False)\n",
    "plt.savefig(output_dir+i+'_doublet_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [4,3]\n",
    "ax = sc.pl.umap(adata_files[i], color = 'doublet_cutoff', groups = ['doublet'],show = False)\n",
    "plt.savefig(output_dir+i+'_doublet2_umap.png', dpi = 100,bbox_inches='tight')\n",
    "plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e00a324c-f3c9-4007-a61f-7e308b3aa128",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata_files[i], 'leiden_post_QC_highres', method = 'wilcoxon')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6495b4ac-9d91-4f9f-a695-07762a43f1ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    adata_files[i],\n",
    "    groups = ['24','25','33','38'],\n",
    "    n_genes=15,\n",
    "    values_to_plot=\"logfoldchanges\", cmap='bwr',\n",
    "    vmin=-4,\n",
    "    vmax=4,\n",
    "    min_logfoldchange=2,\n",
    "    colorbar_title='log fold change'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "383187c2-e92e-4dc2-85ba-e2ed30affbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 'SE8_9_pass_doublet'\n",
    "sc.pl.umap(adata_files[i], color = ['leiden_post_QC_highres','sox10'],legend_loc = 'on data',\n",
    "           cmap = reds, vmin = 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7b66019-b008-43e9-aad6-92a5ec90aef9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy_2025",
   "language": "python",
   "name": "scanpy_2025"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
